{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "models to use:  ['gpt-4']\n"
     ]
    }
   ],
   "source": [
    "import autogen\n",
    "from autogen.agentchat.contrib.retrieve_assistant_agent import RetrieveAssistantAgent\n",
    "from autogen.agentchat.contrib.retrieve_user_proxy_agent import RetrieveUserProxyAgent\n",
    "import chromadb\n",
    "\n",
    "config_list = autogen.config_list_from_json(\n",
    "    env_or_file=\"modelConfig.json\",\n",
    "    file_location=\".\",\n",
    "    filter_dict={\n",
    "        \"model\": {\n",
    "            \"gpt-3.5-turbo-1106\"\n",
    "        }\n",
    "    },\n",
    ")\n",
    "\n",
    "assert len(config_list) > 0\n",
    "print(\"models to use: \", [config_list[i][\"model\"] for i in range(len(config_list))])\n",
    "\n",
    "\n",
    "# 1. create an RetrieveAssistantAgent instance named \"assistant\"\n",
    "bioinformatics_assistant = RetrieveAssistantAgent(\n",
    "    name=\"bioinformatics_agent\", \n",
    "    system_message=\"Hello, you are the bioinformatician. You can help me with analyzing and interpreting biological data. Your skills include statistical analysis, DNA/RNA sequencing analysis, designing bioinformatics workflows, and providing guidance on best practices for computational biology research. Please help me analyze and make sense of large-scale omics data!\",\n",
    "    llm_config={\n",
    "        \"timeout\": 600,\n",
    "        \"config_list\": config_list,\n",
    "    },\n",
    ")\n",
    "\n",
    "bioinformatics_proxyagent = RetrieveUserProxyAgent(\n",
    "    name=\"bioinformatics_proxy_agent\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    max_consecutive_auto_reply=10,\n",
    "    retrieve_config={\n",
    "        \"docs_path\": \"BioinformaticsPDF\",  # change this to your own path, such as https://raw.githubusercontent.com/microsoft/autogen/main/README.md\n",
    "        \"chunk_token_size\": 1000,\n",
    "        \"model\": config_list[0][\"model\"],\n",
    "        \"client\": chromadb.PersistentClient(path=\"chroma_bioinformatician_db\"),\n",
    "        \"embedding_model\": \"all-mpnet-base-v2\",\n",
    "        \"get_or_create\": True,  # set to False if you don't want to reuse an existing collection, but you'll need to remove the collection manually\n",
    "    },\n",
    ")\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T09:31:35.149861Z",
     "start_time": "2023-11-12T09:31:35.139104Z"
    }
   },
   "id": "bb37a9d648002fdb"
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying to create collection.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "max_tokens is too small to fit a single line of text. Breaking this line:\n",
      "\tDesigning your Analysis  ...\n",
      "Failed to split docs with must_break_at_empty_line being True, set to False.\n",
      "Number of requested results 20 is greater than number of elements in index 5, updating n_results = 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doc_ids:  [['doc_0', 'doc_4', 'doc_3', 'doc_2', 'doc_1']]\n",
      "\u001B[32mAdding doc_id doc_0 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_4 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_3 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_2 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_1 to context.\u001B[0m\n",
      "\u001B[33mbioinformatics_proxy_agent\u001B[0m (to bioinformatics_agent):\n",
      "\n",
      "You're a retrieve augmented chatbot. You answer user's questions based on your own knowledge and the\n",
      "context provided by the user. You should follow the following steps to answer a question:\n",
      "Step 1, you estimate the user's intent based on the question and context. The intent can be a code generation task or\n",
      "a question answering task.\n",
      "Step 2, you reply based on the intent.\n",
      "If you can't answer the question with or without the current context, you should reply exactly `UPDATE CONTEXT`.\n",
      "If user's intent is code generation, you must obey the following rules:\n",
      "Rule 1. You MUST NOT install any packages because all the packages needed are already installed.\n",
      "Rule 2. You must follow the formats below to write your code:\n",
      "```language\n",
      "# your code\n",
      "```\n",
      "\n",
      "If user's intent is question answering, you must give as short an answer as possible.\n",
      "\n",
      "User's question is: explain your role.\n",
      "\n",
      "Context is: Designing your Analysis \n",
      "By Tallulah Andrews What is your question? \n",
      "What data do you need to answer it? What progenitor states exist between hematopoietic stem cells and each \n",
      "differentiated blood cell-types? \n",
      "How do different Huntington gene variants affect neuronal identity/function? \n",
      "How many cell-types exist in different regions of the liver? \n",
      "How does the communication between glia and neurons change in Parkinson’s \n",
      "disease? \n",
      "What transcription factor(s) control the differentiation of pancreatic cell-types? \n",
      "Example Questions: Algorithms/Tools will always give you an answer! \n",
      "That doesn’t mean that answer is “true”. Part 1: When and what to \n",
      "normalize/correct? Batch Effects / Sample Integration \n",
      "-Technical Noise (unknown origin) \n",
      "-Can create false clusters/DE genes in our analysis \n",
      "-If modelled correctly, helps assess confidence of our \n",
      "conclusions \n",
      "-Difficult to model because may not be linear / uniform \n",
      "across cell-types Unconfounded Experiments vs Replicates \n",
      "Biological “Noise” aka Confounders \n",
      "Examples: \n",
      "-Cell cycle \n",
      "-Genetic background / individual \n",
      "-Age\n",
      "-Circadian rhythm \n",
      "-Cell stress Solutions: \n",
      "-Regress / correct / normalize it \n",
      "-Exclude cells/genes most \n",
      "affected by it \n",
      "-Include as a covariate in \n",
      "models \n",
      "-Ignore it and hope the biology \n",
      "we are interested in still comes \n",
      "through. \n",
      "The Cell Cycle \n",
      "Regressing out the cell-cycle will also \n",
      "remove all variability that is confounded \n",
      "with it. \n",
      "Development/Differentiation \n",
      "-Cell cycle may be a confounder \n",
      "Mature tissue \n",
      "-Usually cells are not cycling, non-issue \n",
      "Cancer \n",
      "-Cell cycle is often biologically interesting Individual Variation \n",
      "Human / Patient samples \n",
      "Often must be confounded with \n",
      "disease or treatment conditions \n",
      "-3 patients with diabetes vs \n",
      "3 patients without \n",
      "Treated similar to batch effects. Stress Response \n",
      "Cluster:   Prog      Chol        CSC Stress Cancer Stem Metabolic Redox Stress Common when dissociating tissues \n",
      "Highly cell-type dependent \n",
      "May depend on condition / \n",
      "replicate \n",
      "Often non-linear -> Difficult to \n",
      "regress \n",
      "Typically exclude affected cells When to impute/smooth data? - Visualization only \n",
      "Part 2: Clustering vs Pseudotime \n",
      "You shouldn’t be doing both. Part 2: Clustering vs Pseudotime \n",
      "You shouldn’t be doing  publishing both. \n",
      "(usually) Case Study: Malaria Cell Atlas \n",
      "Problem: What is a “good” cluster? \n",
      "How many \n",
      "clusters are there \n",
      "in this data? \n",
      "Pause the video and decide \n",
      "on your answer The Truth Clustering: Common Assumptions \n",
      "-Clusters are roughly the same size \n",
      "-May fail to detect a rare cell-types \n",
      "-Clusters are roughly the same density (almost all) \n",
      "-Heterogeneous/noisy clusters often split up into many smaller clusters \n",
      "-Clusters are distinct (almost all) \n",
      "-Gradients, trajectories, continuums are split up along their length in a way to optimize one of \n",
      "the above qualities. \n",
      "-Resolution parameters Clustering: Common Assumptions \n",
      "-Clusters are roughly the same size \n",
      "-May fail to detect a rare cell-type \n",
      "-May arbitrarily split up large clusters \n",
      "-Clusters are roughly the same density (almost all) \n",
      "-Heterogeneous/noisy clusters often split up into many smaller clusters \n",
      "-Clusters are distinct (almost all) \n",
      "-Gradients, trajectories, continuums are split up along their length in a way to optimize one of \n",
      "the above qualities. \n",
      "-Resolution parameters Clustering: Common Assumptions \n",
      "-Clusters are roughly the same size \n",
      "-May fail to detect a rare cell-type \n",
      "-May arbitrarily split up large clusters \n",
      "-Clusters are roughly the same density  \n",
      "-Noisy / sparse clusters often split up into many smaller clusters \n",
      "-Clusters are distinct (almost all) \n",
      "-Gradients, trajectories, continuums are split up along their length in a way to optimize one of \n",
      "the above qualities. \n",
      "-Resolution parameters What happens if you use a clustering method on a gradient? \n",
      "Pause the video and decide \n",
      "on your answer Gradient 1: \n",
      "Gradient 2: Stem cells Differentiated \n",
      "Apical Basal variability What happens if you use a clustering method on a gradient? \n",
      "Gradient 1: \n",
      "How does the communication between glia and neurons change in Parkinson’s \n",
      "disease? \n",
      "●What type of question is being asked? \n",
      "Comparative & Relational \n",
      "●What data do you need to answer the question? \n",
      "Single-cell RNAseq from both glia and neurons in normal samples and samples affected by \n",
      "Parkinson’s disease. Exercise Answers: (5) \n",
      "What transcription factor(s) control the differentiation of pancreatic cell-types? \n",
      "●What type of question is being asked? \n",
      "Relational & Descriptive \n",
      "●What data do you need to answer the question? \n",
      "Single-cell RNAseq at multiple time points during pancreatic development or in vitro  \n",
      "differentiation \n",
      "○Does cell-type A communicate with cell-type B? \n",
      "○Is cell-type A a progenitor of cell-type B? \n",
      "●Comparative Analyses (multiple conditions): \n",
      "○How does this cell-type or developmental process change during disease? \n",
      "○In response to a drug/mutation? \n",
      "○When in contact with a different cell-type? \n",
      "What is this? Types of questions \n",
      "●Descriptive Analysis (one condition): \n",
      "○How many cell-types are there? What are they? \n",
      "○Is Gene X expressed in a particular cell-type? \n",
      "○Are there developmental trajectories/gradients? \n",
      "○Where is this cell-type located? \n",
      "●Relational (one condition, multiple cell-types): \n",
      "○Which cell-types are similar to other cell-types? \n",
      "○Does cell-type A communicate with cell-type B? \n",
      "○Is cell-type A a progenitor of cell-type B? \n",
      "●Comparative Analyses (multiple conditions): \n",
      "○How does this cell-type or developmental process change during disease? \n",
      "○In response to a drug/mutation? \n",
      "○When in contact with a different cell-type? \n",
      "How are things related \n",
      "to each other? Types of questions \n",
      "●Descriptive Analysis (one condition): \n",
      "○How many cell-types are there? What are they? \n",
      "○Is Gene X expressed in a particular cell-type? \n",
      "○Are there developmental trajectories/gradients? \n",
      "○Where is this cell-type located? \n",
      "●Relational (one condition, multiple cell-types): \n",
      "○Which cell-types are similar to other cell-types? \n",
      "○Does cell-type A communicate with cell-type B? \n",
      "○Is cell-type A a progenitor of cell-type B? \n",
      "●Comparative Analyses (multiple conditions): \n",
      "○How does this cell-type or developmental process change during disease? \n",
      "○In response to a drug/mutation? \n",
      "○When in contact with a different cell-type? \n",
      "How do things change \n",
      "when X happens? <- Tumour Type -> <- Marker Genes -> Types of questions \n",
      "●Descriptive Analysis (one condition): \n",
      "○How many cell-types are there? What are they? \n",
      "○Is Gene X expressed in a particular cell-type? \n",
      "○Are there developmental trajectories/gradients? \n",
      "○Where is this cell-type located? \n",
      "●Relational (one condition, multiple cell-types): \n",
      "○Which cell-types are similar to other cell-types? \n",
      "○Does cell-type A communicate with cell-type B? \n",
      "○Is cell-type A a progenitor of cell-type B? \n",
      "●Comparative Analyses (multiple conditions): \n",
      "○How does this cell-type or developmental process change during disease? \n",
      "○In response to a drug/mutation? \n",
      "○When in contact with a different cell-type? Most current tools exist \n",
      "in this space. \n",
      "Differential \n",
      "Expression Interactive Exercise: Research Questions \n",
      "For each situation: \n",
      "●What type of question is being asked? \n",
      "●What data do you need to answer the question? \n",
      "Pause the video to write down your answers before continuing. (1) What progenitor states exist between hematopoietic stem cells and each \n",
      "differentiated blood cell-types? \n",
      "(2) How do different Huntington gene variants affect neuronal identity/function? \n",
      "(3) How many cell-types exist in different regions of the liver? \n",
      "(4) How does the communication between glia and neurons change in Parkinson’s \n",
      "disease? \n",
      "(5) What transcription factor(s) control the differentiation of pancreatic cell-types? \n",
      "Exercise Situations: \n",
      "Pause the video now to write down your answers \n",
      "(5 minutes allotted) Exercise Answers: (1) \n",
      "What progenitor states exist between hematopoietic stem cells and each \n",
      "differentiated blood cell-types? \n",
      "●What type of question is being asked? \n",
      "Relational & Descriptive \n",
      "●What data do you need to answer the question? \n",
      "Single cell expression for one sample of differentiated blood cells, and from bone marrow \n",
      "and all other sites of blood cell maturation. Exercise Answers: (2) \n",
      "How do different Huntington gene variants affect neuronal identity/function? \n",
      "●What type of question is being asked? \n",
      "Comparative \n",
      "●What data do you need to answer the question? \n",
      "Single cell RNAseq from brain samples or brain organoids from multiple samples carrying \n",
      "different Huntington gene variants. \n",
      "(Bulk RNAseq of sorted neurons is also valid) Exercise Answers: (3) \n",
      "How many cell-types exist in different regions of the liver? \n",
      "●What type of question is being asked? \n",
      "Descriptive (perhaps Comparative) \n",
      "●What data do you need to answer the question? \n",
      "Single cell RNAseq from one or more different regions a liver. Exercise Answers: (4) \n",
      "-Oscillatory patterns will not be detected! \n",
      "-Independent of distribution of data values \n",
      "-Can be applied to batch-corrected, scaled, normalized, regressed data \n",
      "-Do not account for confounding factors contributing to differences in \n",
      "expression What is the “best”? \n",
      "Soneson and Robinson. 2018. Nature Methods. 15 : 255-261 doi: 10.1038/nmeth.4612 Conclusions: \n",
      "•Seurat performs poorly \n",
      "•Both GLM and \n",
      "non-parametric tests work \n",
      "well when used correctly \n",
      "•MAST is best GLM \n",
      "•Wilcoxon  is best \n",
      "nonparametric \n",
      "DE across multiple batches and multiple conditions \n",
      "●Subset your data to a single cell-type, then perform a DE test: \n",
      "●General Linear Model \n",
      "○Include biological condition(s) of interest as a predictor. \n",
      "○Include batch ID as a random effect. \n",
      "○MAST or NEBULA \n",
      "●Pseudobulks \n",
      "○Sum raw counts of cells in each batch x condition. These are “pseudobulk” samples. \n",
      "○Use existing bulk RNAseq DE tools on the pseudobulks (e.g. edgeR). Analysis of complex experiments \n",
      "How would you test the following questions? (Hint: there is more than one good answer) \n",
      "1)Effect of stimulation on gene expression in three batches of T-cells, each \n",
      "batch contains 50% stimulated and 50% unstimulated. \n",
      "2)Cell-type specific effects of diabetes in the pancreas, three replicates were \n",
      "performed for diabetic and non-diabetic pancreas samples. \n",
      "Pause the video now to write down your answers \n",
      "(15 minutes allotted) Effect of stimulation on gene expression in three batches of T-cells, \n",
      "each batch contains 50% stimulated and 50% unstimulated. \n",
      " Stimulated Unstimulated \n",
      "Batch 1 150 150\n",
      "Batch 2 150 150\n",
      "Batch 3 150 150Example Analysis: \n",
      "1.Assign cells from each batch to \n",
      "original sample \n",
      "2.Batch Correction \n",
      "3.Clustering & Marker detection \n",
      "(identify sub-types of T-cells) \n",
      "4.Differential expression with \n",
      "Wilcox-rank-sum test for each \n",
      "cluster before/after stimulation Effect of stimulation on gene expression in three batches of T-cells, \n",
      "each batch contains 50% stimulated and 50% unstimulated. \n",
      " Stimulated Unstimulated \n",
      "Batch 1 150 150\n",
      "Batch 2 150 150\n",
      "Batch 3 150 150Analysis Option 2: \n",
      "1.Assign cells from each batch to \n",
      "original sample \n",
      "2.Clustering & Marker detection \n",
      "(identify sub-types of T-cells) \n",
      "3.Differential expression with a \n",
      "GLM including Batch and \n",
      "Stimulation as predictors. Cell-type specific effects of diabetes in the pancreas, three replicates \n",
      "were performed for diabetic and non-diabetic pancreas samples. \n",
      " Diabetic Healthy \n",
      "Donor 1 1500 0\n",
      "Donor 2 1500 0\n",
      "Donor 3 1500 0\n",
      "Donor 4 0 1500 \n",
      "Donor 5 0 1500 \n",
      "Donor 6 0 1500 Example Analysis: \n",
      "1.Cluster each sample separately \n",
      "2.Match clusters across donors \n",
      "3.Use a GLM with Donor as a \n",
      "random effect, and \n",
      "Diabetic/Healthy as the predictor Cell-type specific effects of diabetes in the pancreas, three replicates \n",
      "were performed for diabetic and non-diabetic pancreas samples. \n",
      " Diabetic Healthy \n",
      "Donor 1 1500 0\n",
      "Donor 2 1500 0\n",
      "Donor 3 1500 0\n",
      "Donor 4 0 1500 \n",
      "Donor 5 0 1500 \n",
      "Donor 6 0 1500 Analysis Option 2: \n",
      "1.Integrate samples together \n",
      "without correcting batch effects \n",
      "2.Cluster on Integrated data \n",
      "3.Create mini-bulks by summing \n",
      "raw counts for cells from each \n",
      "donor in each cluster. \n",
      "4.Perform traditional bulk-RNAseq \n",
      "DE within each cell-type. END Types of questions \n",
      "●Descriptive Analysis (one condition): \n",
      "○How many cell-types are there? What are they? \n",
      "○Is Gene X expressed in a particular cell-type? \n",
      "○Are there developmental trajectories/gradients? \n",
      "○Where is this cell-type located? \n",
      "●Relational (one condition, multiple cell-types): \n",
      "○Which cell-types are similar to other cell-types? \n",
      "Gradient 2: \n",
      "Trajectories are split up along their length in a way to \n",
      "optimize uniformity of clusters. Clustering: Common Assumptions \n",
      "-Clusters are roughly the same size \n",
      "-May have difficulty detecting a rare cell-type \n",
      "-May arbitrarily split up large clusters \n",
      "-Clusters are roughly the same density \n",
      "-Noisy / sparse clusters often split up into many smaller clusters \n",
      "-Clusters exist at a fixed & predetermined resolution Clustering the same data with different parameters: \n",
      "Seurat Applied to MCA data: \n",
      "Morphologically there are 3 distinct stages \n",
      "of parasites, using marker genes we could \n",
      "link these clusters to them. \n",
      "Are the subdivisions actually meaningful? \n",
      "Or is this a smooth continuum that has \n",
      "been arbitrarily divided up by a clustering \n",
      "algorithm? Clustering: How do you know a cluster is “real”? \n",
      "1.Robustness to clustering method/parameters \n",
      "2.Significant marker genes / differentially expressed genes \n",
      "3.Known marker genes \n",
      "4.Quality statistics: \n",
      "e.g. Silhouette Index \n",
      "5.Consistency across experimental replicates or with reference data \n",
      "e.g. scmap \n",
      "6.Spatial structure \n",
      "7.Experimental validation Seeing that they “look good” in a visualization \n",
      "is not a good way to validate clustering Malaria Cell Atlas Data: \n",
      "1.Robustness to clustering method/parameters \n",
      "2.Significant marker genes / differentially expressed genes \n",
      "3.Known marker genes \n",
      "4.Quality statistics: \n",
      "a.e.g. Silhouette Index \n",
      "5.Consistency across experimental replicates or with reference data \n",
      "a.e.g. scmap \n",
      "6.Spatial structure \n",
      "7.Experimental validation \n",
      "NA\n",
      "NAPseudotime: Common Assumptions \n",
      "-All cells exist on a smooth continuum \n",
      "-Cells belonging to a distinct cluster create a false branch \n",
      "-Varying thickness (cell density) may be mistaken for a branch \n",
      "Pseudotime: Common Assumptions \n",
      "-All cells exist on a smooth continuum \n",
      "-Cells belonging to a distinct cluster create a false branch \n",
      "-Varying thickness (cell density) may be mistaken for a branch \n",
      "-The continuum is a line/curve/tree and has a start and an end \n",
      "-Cycles fail to be properly modelled \n",
      "-Multiple overlapping gradiants create “shapes” (e.g. triangle) which may be forced \n",
      "into a “Y” shaped topology \n",
      "-If you specified the number of branches/clusters the tool will find that number even \n",
      "if that is not correct. Monocle Applied to MCA data: What happened? \n",
      "PCA Monocle \n",
      "Pause the video and decide \n",
      "on your answer Monocle Applied to MCA data: \n",
      "Errors due to model assumptions: \n",
      "1.Arbitrarily cut the circle open \n",
      "2.  False branches at most heavily \n",
      "sampled portion of the cycle \n",
      "Danger! \n",
      "We could have interpreted \n",
      "one/both of the branches as cells \n",
      "moving in/out of the IDC \n",
      "Malaria Cell Atlas - what did we do? (both) Why did we end up doing both? \n",
      "Part 3: Visualization \n",
      "All models are wrong, some are useful  -George Box \n",
      "Visualizations Visualization: Box Analogy \n",
      "PC 1 PC 2 \n",
      "PC 1 PC 3 PC1\n",
      "PC2PC3Visualization: Box Analogy \n",
      "UMAP 1 UMAP 2 Visualization: What do you choose to show? \n",
      "Accurate distances between points. - PCA \n",
      "Overall structure - UMAP \n",
      "Distinct Clusters - t-SNE \n",
      "Trajectories/gradiants - Diffusion Map How many clusters can you see? \n",
      "tSNE UMAP Part 4: Differential Expression General Linear Models: Common Assumptions \n",
      "-Changes are Linear \n",
      "-Oscillatory patterns will not be detected! \n",
      " \n",
      "Pseudotime Pseudotime General Linear Models: Common Assumptions \n",
      "-Changes are Linear \n",
      "-Oscillatory patterns will not be detected! \n",
      "-Noise follows a specific distribution \n",
      "-Negative Binomial - Suitable for UMI  counts \n",
      "-Zero Inflated Negative Binomial - Suitable for UMI counts or Read counts General Linear Models: Common Assumptions \n",
      "-Changes are Linear \n",
      "-Oscillatory patterns will not be detected! \n",
      "-Noise follows a specific distribution \n",
      "-Negative Binomial - Suitable for UMI  counts \n",
      "-Zero Inflated Negative Binomial - Suitable for UMI counts or Read counts \n",
      " \n",
      "-Changes in Expression caused by different factors are Additive \n",
      "-i.e. batch effects are the same across cell-types Non-parametric Tests \n",
      "-Changes are Monotonic \n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mbioinformatics_agent\u001B[0m (to bioinformatics_proxy_agent):\n",
      "\n",
      "As a bioinformatician, my role is to assist you with the analysis and interpretation of large-scale biological data such as genomics. This includes handling raw sequencing data generated by next-generation sequencing platforms, designing data analysis workflows, and applying appropriate statistical methods to decipher patterns.\n",
      "\n",
      "In a specific context like the one you presented, I can help answer questions like \"What progenitor states exist between hematopoietic stem cells and each differentiated blood cell-types?\" or \"How does the communication between glia and neurons change in Parkinson’s disease?\" by applying computational methods to process, analyze and interpret the omics data.\n",
      "\n",
      "Furthermore, I would help in interpreting the results of various bioinformatics tools and algorithms, understanding their implications, figures such as PCA, UMAP, and t-SNE plots, and biologically relevant conclusions that can be drawn from those.\n",
      "\n",
      "Workflows I design might include steps like normalization/correction of data, application of clustering or pseudotime analysis, visualization of results, and identification of differentially expressed genes. In each of these steps, careful attention would be given to meet the assumptions of the analysing methods and avoid false interpretations.\n",
      "\n",
      "Overall, my role involves transforming raw, complex data into meaningful biological insights which can further be used for deeper research or direct application.\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "query = \"explain your role.\"\n",
    "\n",
    "bioinformatics_proxyagent.initiate_chat(bioinformatics_assistant, problem=query)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T09:31:40.625323Z",
     "start_time": "2023-11-12T09:31:38.427637Z"
    }
   },
   "id": "702098887a51ffbd"
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying to create collection.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "max_tokens is too small to fit a single line of text. Breaking this line:\n",
      "\tA distal lung organoid model to study interstitial ...\n",
      "Failed to split docs with must_break_at_empty_line being True, set to False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doc_ids:  [['doc_19', 'doc_22', 'doc_21', 'doc_20', 'doc_28', 'doc_25', 'doc_27', 'doc_18', 'doc_2', 'doc_17', 'doc_26', 'doc_16', 'doc_15', 'doc_42', 'doc_44', 'doc_13', 'doc_40', 'doc_46', 'doc_12', 'doc_7']]\n",
      "\u001B[32mAdding doc_id doc_19 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_22 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_21 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_20 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_28 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_25 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_27 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_18 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_2 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_17 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_26 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_16 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_15 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_42 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_44 to context.\u001B[0m\n",
      "\u001B[33mexperimentalist_proxy_agent\u001B[0m (to experimentalist_agent):\n",
      "\n",
      "You're a retrieve augmented chatbot. You answer user's questions based on your own knowledge and the\n",
      "context provided by the user. You should follow the following steps to answer a question:\n",
      "Step 1, you estimate the user's intent based on the question and context. The intent can be a code generation task or\n",
      "a question answering task.\n",
      "Step 2, you reply based on the intent.\n",
      "If you can't answer the question with or without the current context, you should reply exactly `UPDATE CONTEXT`.\n",
      "If user's intent is code generation, you must obey the following rules:\n",
      "Rule 1. You MUST NOT install any packages because all the packages needed are already installed.\n",
      "Rule 2. You must follow the formats below to write your code:\n",
      "```language\n",
      "# your code\n",
      "```\n",
      "\n",
      "If user's intent is question answering, you must give as short an answer as possible.\n",
      "\n",
      "User's question is: explain your role.\n",
      "\n",
      "Context is: 0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "0123SCGB3A2d25 d40 d60 d80 d25 d40 d60 d807\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "592\n",
      "0 0 0 006 d25 d40 d60 d80\n",
      "Umap_1Umap_2d40d60d80\n",
      "FeaturesFOXA1 NKX2.1CPMSFTPB ABCA3 NAPSANAPSASLC43A2LPCAT1SCGB3A2CDX2GATA4THY1\n",
      "COL1A2ZEB1 TBX4WNT2BKi67\n",
      "ACTA2Identity\n",
      "d25ab\n",
      "c%\n",
      "0\n",
      "20\n",
      "40\n",
      "60\n",
      "80\n",
      "Expre-\n",
      "ssion\n",
      "1.0\n",
      "0.50\n",
      "–0.5–1.0\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –8\n",
      "03\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "–66\n",
      "04 –4 –803\n",
      "–3\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345MKI67\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345ACTA2d25\n",
      "d40 d60d80\n",
      "d25\n",
      "d40d60d80a\n",
      "b–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "–10 –5–10–5\n",
      "7713\n",
      "3\n",
      "8\n",
      "10\n",
      "16\n",
      "4\n",
      "9155\n",
      "2141710\n",
      "613\n",
      "1111 3\n",
      "8\n",
      "711\n",
      "14\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345CPM\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345SFTPB\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345NAPSA\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345ABCA3\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015–10 –5\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345SLC34A2\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345ZEB1\n",
      "83 11\n",
      "1013\n",
      "6\n",
      "54\n",
      "712\n",
      "14\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345COL1A2\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345TBX4\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345WNT2B\n",
      "83 11\n",
      "1061312\n",
      "54\n",
      "71416\n",
      "9150\n",
      "21\n",
      "30\n",
      "1\n",
      "46\n",
      "589\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "126\n",
      "07\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "0123WNT2B\n",
      "7\n",
      "30\n",
      "416\n",
      "59\n",
      "2\n",
      "8 06\n",
      "5\n",
      "1\n",
      "42\n",
      "379\n",
      "8\n",
      "126\n",
      "70\n",
      "3 459\n",
      "80\n",
      "312\n",
      "476\n",
      "589\n",
      "01234Ki6770\n",
      "31\n",
      "48\n",
      "529 6\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "57\n",
      "486 9\n",
      "0123ACTA2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "BMP4\n",
      "Retinoic acid\n",
      "Fig. 1 | Overview of the protocol. a, Schematic representation of the distinct stages of the protocol.\n",
      "b, Representative bright ﬁeld images at indicated days of the protocol. Scale bars, 200 μm.PROTOCOL NATURE PROTOCOLS\n",
      "2284 NATURE PROTOCOLS | VOL 18 | JULY 2023 | 2283 –2312 | www.nature.com/nprotApplications of the method\n",
      "The most unique feature of the organoids described here is the presence of mesenchyme that allows\n",
      "studies of epithelial –mesenchymal interactions and therefore recapitulation and modeling of at least\n",
      "some ILDs. ILDs are a heterogeneous group of entities with diverse or unknown etiologies affectingthe lung interstitium\n",
      "33. Development of pulmonary ﬁbrosis (PF), most commonly associated with\n",
      "idiopathic pulmonary ﬁbrosis (IPF), portends the worst prognosis33,34. Injury to AT2 cells probably\n",
      "plays a critical role in the development of both familial and sporadic IPF, as mutations causingfamilial IPF speci ﬁcally affect AT2 function, while depletion of AT2 cells in mouse models is asso-\n",
      "ciated with ﬁbrosis\n",
      "35. In addition, a fraction of patients with familial IPF have heterozygous muta-\n",
      "tions in telomerase components or other genes involved in telomere maintenance36–43. Furthermore,\n",
      "several susceptibility loci have been identi ﬁed that affect telomere length44. As telomeropathy affects\n",
      "stem cell function and AT2 cells are facultative distal lung stem cells4,45, these associations suggest a\n",
      "role for AT2 cells.\n",
      "Finding therapeutic targets and novel drug treatments requires models to dissect pathogenetic\n",
      "mechanisms, identify biomarkers, discover drug targets and perform drug testing. It has been chal-lenging to model ILDs in animal models\n",
      "34,35,46–48. A complementary, more malleable, but also more\n",
      "reductionist avenue are lung organoids derived from hPSCs. We showed that PF caused by mutation\n",
      "in Hermansky –Pudlak syndrome (HPS) genes could be modeled28. HPS is a genetic syndrome caused\n",
      "by abnormal biogenesis and traf ﬁcking of lysosome-related organelles and characterized by hypo-\n",
      "pigmentation, bleeding diathesis and in some but not all genotypes, PF, also known as HPS-associatedinterstitial pneumonia (HPSIP). Fibrosis in HPSIP is most probably linked to the fact that home-ostasis of LBs in AT2 cells is abnormal and surfactant secretion is reduced, as LBs are also lysosome-related organelles\n",
      "49. Of the nine genotypes, only three (HPS1, 2 and 4) are clinically associated with\n",
      "HPSIP. Accordingly, we found that only organoids derived from ESCs with mutation in HPS1, 2 or4 showed ﬁbrotic changes, whereas organoids with a mutation in HPS8, which is clinically not\n",
      "associated with ﬁbrosis, did not\n",
      "28. This model therefore recapitulates the clinical incidence of PF in\n",
      "HPS. By comparing genome-wide expression data from epithelial cells isolated from lung organoids\n",
      "generated from the ﬁbrosis prone (HPS1−/−and HPS2−/−) and non- ﬁbrosis-prone (HPS8−/−)\n",
      "mutants, we furthermore identi ﬁed interleukin-11 as a therapeutic target critical to ﬁbrosis. This\n",
      "provides proof-of-principle evidence that pathogenetic mechanism and therapeutic targets can beidenti ﬁed using lung organoids with mutations associated with diseases\n",
      "28.\n",
      "The organoids can also be used to model viral infection. They can be infected by respiratory\n",
      "viruses, such as respiratory syncytial virus, parain ﬂuenza and measles. We showed that the organoids\n",
      "are authentic infection models that do not promote the emergence of culture-adapted variants in thepropagation of parain ﬂuenza\n",
      "27. In the case of respiratory syncytial virus, we could show that infected\n",
      "epithelial cells are dislodged into the lumen, as is observed in bronchiolitis caused by this virus49.\n",
      "The organoids can also be used for modeling human lung development. As an example, single-cell\n",
      "RNA sequencing (scRNAseq) and confocal analysis is presented in the ‘Anticipated results ’section,\n",
      "showing progressive loss of other endodermal fates present early in organoid development, and\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "0123THY1\n",
      "70\n",
      "31\n",
      "46\n",
      "8\n",
      "529\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "01234COL1A2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "0123ZEB1\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "00.51.01.52.0TBX4\n",
      "7\n",
      "54\n",
      "7916\n",
      "15140\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345LPCAT1\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "915160\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345CDX2\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345SCGB3A2\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345GATA4\n",
      "83 11\n",
      "1013\n",
      "612\n",
      "54\n",
      "714\n",
      "916\n",
      "150\n",
      "21\n",
      "17051015\n",
      "05 1 0\n",
      "UMAP_1UMAP_212345THY1\n",
      "83 11\n",
      "1013\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "126\n",
      "078\n",
      "4 359\n",
      "12\n",
      "5 0\n",
      "376\n",
      "489\n",
      "0123SLC34A2\n",
      "7\n",
      "30\n",
      "416\n",
      "52\n",
      "89 6\n",
      "05\n",
      "42\n",
      "8\n",
      "1\n",
      "379\n",
      "26\n",
      "10\n",
      "4 3879\n",
      "5\n",
      "12\n",
      "0\n",
      "3 4786 9\n",
      "5\n",
      "0123LPCAT17\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "182\n",
      "4 379\n",
      "1026\n",
      "7\n",
      "38\n",
      "459\n",
      "0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "00.40.81.21.6GATA4307\n",
      "1\n",
      "46\n",
      "89\n",
      "52\n",
      "06\n",
      "18\n",
      "42\n",
      "3579\n",
      "1026\n",
      "7\n",
      "38\n",
      "49\n",
      "5 0\n",
      "312\n",
      "5\n",
      "4786 9\n",
      "0123CDX2\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "42\n",
      "3078\n",
      "19\n",
      "26\n",
      "70\n",
      "31\n",
      "49\n",
      "582\n",
      "4106\n",
      "7\n",
      "359\n",
      "8\n",
      "0123NKX2.1\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "58\n",
      "42\n",
      "1\n",
      "379\n",
      "26\n",
      "1708\n",
      "3 4592\n",
      "5 0\n",
      "37\n",
      "1\n",
      "46\n",
      "89\n",
      "01234SFTPB\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "026\n",
      "17\n",
      "38\n",
      "459\n",
      "12\n",
      "0\n",
      "3 476\n",
      "859\n",
      "0123ABCA3\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "2\n",
      "06\n",
      "5\n",
      "18\n",
      "42\n",
      "379\n",
      "126\n",
      "07\n",
      "4 35892\n",
      "0\n",
      "37\n",
      "4186\n",
      "59\n",
      "0123NAPSA\n",
      "7\n",
      "30\n",
      "1\n",
      "46\n",
      "8\n",
      "59\n",
      "Physiology 32, 246 –260 (2017).\n",
      "4. Basil, M. C. et al. The cellular and physiological basis for lung repair and regeneration: past, present, and\n",
      "future. Cell Stem Cell 26, 482 –502 (2020).\n",
      "5. Herriges, M. & Morrisey, E. E. Lung development: orchestrating the generation and regeneration of a\n",
      "complex organ. Development 141, 502 –513 (2014).\n",
      "6. Basil, M. C. et al. Human distal airways contain a multipotent secretory cell that can regenerate alveoli.\n",
      "Nature 604, 120 –126 (2022).\n",
      "7. Habermann, A. C. et al. Single-cell RNA sequencing reveals pro ﬁbrotic roles of distinct epithelial and\n",
      "mesenchymal lineages in pulmonary ﬁbrosis. Sci. Adv. 6, eaba1972 (2020).\n",
      "8. Kadur Lakshminarasimha Murthy, P. et al. Human distal lung maps and lineage hierarchies reveal a bipotent\n",
      "progenitor. Nature 604, 111 –119 (2022).\n",
      "9. Morrisey, E. E. & Hogan, B. L. Preparing for the ﬁrst breath: genetic and cellular mechanisms in lung\n",
      "development. Dev. Cell 18,8–23 (2010).\n",
      "10. Burri, P. H. Fetal and postnatal development of the lung. Annu. Rev. Physiol. 46, 617 –628 (1984).\n",
      "11. Nikoli ć, M. Z., Sun, D. & Rawlins, E. L. Human lung development: recent progress and new challenges.\n",
      "Development 145, dev163485 (2018).\n",
      "12. Herring, M. J., Putney, L. F., Wyatt, G., Finkbeiner, W. E. & Hyde, D. M. Growth of alveoli during postnatal\n",
      "development in humans based on stereological estimation. Am. J. Physiol. Lung Cell. Mol. Physiol. 307,\n",
      "L338 –L344 (2014).\n",
      "13. Dye, B. R. et al. A bioengineered niche promotes in vivo engraftment and maturation of pluripotent stem cell\n",
      "derived human lung organoids. eLife 5, e19732 (2016).\n",
      "14. Dye, B. R. et al. In vitro generation of human pluripotent stem cell derived lung organoids. eLife 4, e05098\n",
      "(2015).\n",
      "15. Nikolic, M. Z. et al. Human embryonic lung epithelial tips are multipotent progenitors that can be expanded\n",
      "in vitro as long-term self-renewing organoids. eLife 6, e26575 (2017).\n",
      "16. Dekkers, J. F. et al. Characterizing responses to CFTR-modulating drugs using rectal organoids derived from\n",
      "subjects with cystic ﬁbrosis. Sci. Transl. Med. 8, 344ra84 (2016).\n",
      "17. Salahudeen, A. A. et al. Progenitor identi ﬁcation and SARS-CoV-2 infection in human distal lung organoids.\n",
      "Nature 588, 670 –675 (2020).\n",
      "18. Lamers, M. M. et al. An organoid-derived bronchioalveolar model for SARS-CoV-2 infection of human\n",
      "alveolar type II-like cells. EMBO J. 40, e105912 (2021).\n",
      "19. Jacob, A. et al. Differentiation of human pluripotent stem cells into functional lung alveolar epithelial cells.\n",
      "Cell Stem Cell 21, 472 –488.e10 (2017).\n",
      "20. Chen, Y. W. et al. A three-dimensional model of human lung development and disease from pluripotent stem\n",
      "cells. Nat. Cell Biol. 19, 542 –549 (2017).\n",
      "21. Carvalho, A. L. R. T. de et al. Glycogen synthase kinase 3 induces multilineage maturation of human\n",
      "pluripotent stem cell-derived lung progenitors in 3D culture. Development 146, dev171652 (2019).PROTOCOL NATURE PROTOCOLS\n",
      "2310 NATURE PROTOCOLS | VOL 18 | JULY 2023 | 2283 –2312 | www.nature.com/nprot22. Miller, A. J. et al. Generation of lung organoids from human pluripotent stem cells in vitro. Nat. Protoc. 14,\n",
      "shortening in a family with monoallelic and biallelic PARN pathogenic variants. Hum. Mutat. 40, 2414 –2429\n",
      "(2019).\n",
      "43. Stuart, B. D. et al. Exome sequencing links mutations in PARN and RTEL1 with familial pulmonary ﬁbrosis\n",
      "and telomere shortening. Nat. Genet. 47, 512 –517 (2015).\n",
      "44. Petrovski, S. et al. An exome sequencing study to assess the role of rare genetic variation in pulmonary\n",
      "ﬁbrosis. Am. J. Respir. Crit. Care Med. 196,8 2 –93 (2017).\n",
      "45. Barkauskas, C. E. et al. Type 2 alveolar cells are stem cells in adult lung. J. Clin. Invest. https://doi.org/10.\n",
      "1172/JCI68782 (2013).\n",
      "46. Moeller, A., Ask, K., Warburton, D., Gauldie, J. & Kolb, M. The bleomycin animal model: a useful tool to\n",
      "investigate treatment options for idiopathic pulmonary ﬁbrosis? Int. J. Biochem. Cell Biol. 40, 362 –382 (2008).\n",
      "47. Matute-Bello, G., Frevert, C. W. & Martin, T. R. Animal models of acute lung injury. Am. J. Physiol. Lung Cell\n",
      "Mol. Physiol. 295, L379 –L399 (2008).\n",
      "48. Mulugeta, S., Nureki, S. & Beers, M. F. Lost after translation: insights from pulmonary surfactant for\n",
      "understanding the role of alveolar epithelial dysfunction and cellular quality control in ﬁbrotic lung disease.\n",
      "Am. J. Physiol. Lung Cell Mol. Physiol. 309, L507 –L525 (2015).\n",
      "49. Florin, T. A., Plint, A. C. & Zorc, J. J. Viral bronchiolitis. Lancet 389, 211 –224 (2017).\n",
      "50. Firth, A. L. et al. Generation of multiciliated cells in functional airway epithelia from human induced\n",
      "pluripotent stem cells. Proc. Natl Acad. Sci. USA 111, E1723 –E1730 (2014).\n",
      "51. Yamamoto, Y. et al. Long-term expansion of alveolar stem cells derived from human iPS cells in organoids.\n",
      "Nat. Methods 14, 1097 –1106 (2017).\n",
      "52. Goss, A. M. et al. Wnt2/2b and beta-catenin signaling are necessary and suf ﬁcient to specify lung progenitors\n",
      "in the foregut. Dev. Cell 17, 290 –298 (2009).\n",
      "53. Chapman, D. L. et al. Expression of the T-box family genes, Tbx1-Tbx5, during early mouse development.\n",
      "Dev. Dyn. 206, 379 –390 (1996).\n",
      "54. McCauley, K. B. et al. Ef ﬁcient derivation of functional human airway epithelium from pluripotent stem cells\n",
      "via temporal regulation of Wnt signaling. Cell Stem Cell 20, 844 –857.e6 (2017).NATURE PROTOCOLS PROTOCOL\n",
      "NATURE PROTOCOLS | VOL 18 | JULY 2023 | 2283 –2312 | www.nature.com/nprot 231155. Hawkins, F. et al. Prospective isolation of NKX2-1 –expressing human lung progenitors derived from plur-\n",
      "ipotent stem cells. J. Clin. Invest. 127, 2277 –2294 (2017).\n",
      "56. Jacob, A. et al. Derivation of self-renewing lung alveolar epithelial type II cells from human pluripotent stem\n",
      "cells. Nat. Protoc. 14, 3303 –3332 (2019).\n",
      "57. Rodrigues Toste de Carvalho, A. L. et al. The in vitro multilineage differentiation and maturation of lung and\n",
      "airway cells from human pluripotent stem cell-derived lung progenitors in 3D. Nat. Protoc. 16, 1802 –1829\n",
      "(2021).\n",
      "58. Kanagaki, S. et al. Directed induction of alveolar type I cells derived from pluripotent stem cells via Wnt\n",
      "signaling inhibition. Stem Cells 39, 156 –169 (2021).\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mexperimentalist_agent\u001B[0m (to experimentalist_proxy_agent):\n",
      "\n",
      "As an experimentalist, my abilities involve designing and conducting hands-on laboratory experiments. This could include organizing and setting up lab equipment, developing and executing strict experiment procedures, and collecting important data from those procedures. In addition, I can also handle lab maintenance, ensuring the spaces and tools used are kept clean, safe, and operational. This also involves managing chemicals and other research materials in a safe and organized manner. Yet, your request doesn't seem to pertain any specific field of experiments or lab work. Could you please provide more details on what you need help with?\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# 1. create an RetrieveAssistantAgent instance named \"assistant\"\n",
    "experimentalist_assistant = RetrieveAssistantAgent(\n",
    "    name=\"experimentalist_agent\", \n",
    "    system_message=\"Hello! You are the experimentalist. Your skills are in designing and conducting hands-on lab experiments, including following protocols, collecting and recording data, lab maintenance, and safe chemical/equipment handling. Please help me brainstorming experiments, setting up lab equipment, following detailed protocols, tracking measurements and observations, or assistance with other in-lab experimental work.\",\n",
    "    llm_config={\n",
    "        \"timeout\": 600,\n",
    "        \"config_list\": config_list,\n",
    "    },\n",
    ")\n",
    "\n",
    "experimentalist_proxyagent = RetrieveUserProxyAgent(\n",
    "    name=\"experimentalist_proxy_agent\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    max_consecutive_auto_reply=10,\n",
    "    retrieve_config={\n",
    "        \"docs_path\": \"ExperimentalistPDF\",  # change this to your own path, such as https://raw.githubusercontent.com/microsoft/autogen/main/README.md\n",
    "        \"chunk_token_size\": 1000,\n",
    "        \"model\": config_list[0][\"model\"],\n",
    "        \"client\": chromadb.PersistentClient(path=\"chroma_experimentalist_db\"),\n",
    "        \"embedding_model\": \"all-mpnet-base-v2\",\n",
    "        \"get_or_create\": True,  # set to False if you don't want to reuse an existing collection, but you'll need to remove the collection manually\n",
    "    },\n",
    ")\n",
    "\n",
    "query = \"explain your role.\"\n",
    "\n",
    "experimentalist_proxyagent.initiate_chat(experimentalist_assistant, problem=query)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T09:36:17.198329Z",
     "start_time": "2023-11-12T09:35:13.850280Z"
    }
   },
   "id": "fdd4f232cd629b90"
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying to create collection.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "max_tokens is too small to fit a single line of text. Breaking this line:\n",
      "\tGetting the most of your microscopy data with high-content image analysisBeth Cimini, PhDSenior Grou ...\n",
      "Failed to split docs with must_break_at_empty_line being True, set to False.\n",
      "Number of requested results 20 is greater than number of elements in index 5, updating n_results = 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doc_ids:  [['doc_0', 'doc_1', 'doc_3', 'doc_4', 'doc_2']]\n",
      "\u001B[32mAdding doc_id doc_0 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_1 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_3 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_4 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_2 to context.\u001B[0m\n",
      "\u001B[33mimage_analyst_proxy_agent\u001B[0m (to image_analyst_agent):\n",
      "\n",
      "You're a retrieve augmented chatbot. You answer user's questions based on your own knowledge and the\n",
      "context provided by the user. You should follow the following steps to answer a question:\n",
      "Step 1, you estimate the user's intent based on the question and context. The intent can be a code generation task or\n",
      "a question answering task.\n",
      "Step 2, you reply based on the intent.\n",
      "If you can't answer the question with or without the current context, you should reply exactly `UPDATE CONTEXT`.\n",
      "If user's intent is code generation, you must obey the following rules:\n",
      "Rule 1. You MUST NOT install any packages because all the packages needed are already installed.\n",
      "Rule 2. You must follow the formats below to write your code:\n",
      "```language\n",
      "# your code\n",
      "```\n",
      "\n",
      "If user's intent is question answering, you must give as short an answer as possible.\n",
      "\n",
      "User's question is: explain your role.\n",
      "\n",
      "Context is: Getting the most of your microscopy data with high-content image analysisBeth Cimini, PhDSenior Group LeaderImaging Platform, Broad Institute, Cambridge MA USAImage analysis used to be much less technical\n",
      "https://commons.wikimedia.org/wiki/File:PurkinjeCell.jpg\n",
      "Drawing of Purkinje cells (A) and granule cells (B) from pigeon cerebellum by Santiago Ramón y Cajal, 1899. Instituto Santiago Ramón y Cajal, Madrid, SpainDigital images are ultimately just arrays\n",
      "Digital images are ultimately just arrays\n",
      "Human brains are not built for quantification\n",
      "ØyvindKolås\n",
      "Human brains are not built for quantification\n",
      "ØyvindKolåsHuman brains are not built for quantification\n",
      "ØyvindKolåsDifferent ways of thinking about images\n",
      "Morphological features!Cells!\n",
      "Sample 1!Sample 2!Sample 3!1 Measure/score known phenotypes    2 Profile to characterize samples\n",
      "Free and open-source; Windows, Mac, LinuxCited in 1,500+ papers per yearUsed in 7/10 top pharma companiesIn the Top 10 most popular papers in Genome BiologyRanked most flexible and usable in independent analysis (Wiesmannet al.)\n",
      "Anne CarpenterRay JonesLee KamentskyAllen GoodmanClaire McQuinBeth CiminiDavid StirlingAlice LucasNodarGogoberidzeSoftware overviewImage analysis &quantificationImage-centric data analysis &machine learning\n",
      "Software overviewMeasure everythingAsk question later\n",
      "The CellProfiler interface\n",
      "Pipeline panelSettings panel\n",
      "Module helpStart test modeSet output folderStart analysis runThe CellProfiler interfaceThe CellProfilerinterfaceThe next module to run-Will not execute-Will execute-Will pause during “Run”-Won’t pause during “Run”-Won’t show display-Will show display-Module set correctly-Module has an error-Module giving a warning (such as “won’t run in test mode”)\n",
      "Run until you hit a pauseLeave test modeRun just the next moduleStart over on the next image setLaunch the workspace viewerAdd, subtract, or reorder modulesThe CellProfiler interfaceSet what feeds into and out from every moduleModule categories●File processing:Image input, file output\n",
      "●Image processing:Often used for pre-processing prior to object identification\n",
      "●Object processing:Identification, modification of objects of interest\n",
      "●Measurement:Collection of measurements from objects of interest\n",
      "●Data Tools:Measurement exploration, measurement output\n",
      "●Advanced:Typicallymodules for 3D analyses\n",
      "●Worm Toolbox: C. elegans-specific operationsSearch modules for keywordsCellProfiler figure windows\n",
      "●The figure window has additional menu options \n",
      "●Toolbar menu: Home, pan, zoom in/out\n",
      "●CellProfilerImage Tools○Show pixel data (location, intensity)○Measure length between any two points just by clicking and dragging\n",
      "Input Modules●4 modules in total, handle all the “bookkeeping” of what your experimental setup is○Images (mandatory) –tell CellProfilerwhich images you want to analyze○Metadata (optional if one field of view per file) –give CellProfilermetadata from the file header OR file name○NamesAndTypes(mandatory) –tell CellProfilerif 2D vs 3D, how to break down channels, any other bookkeeping○Groups (optional outside of tracking, Z projection, or whole-plate correction pipelines) –tell CellProfilerif it is important to keep any image sets together during processing●Unlike the rest of CellProfiler, these 4 modules are NOT linear, they work together in concert-so if Groups is on and is set incorrectly, NamesAndTypeswill break●Seeagreatblog post aboutthis,withlinks to a video tutorial, at broad.io/CellProfilerInputInput Modules -ImagesDrop images here (or right click to search)Optionally filter by text or file typeInput Modules -MetadataKeep this off if not usingMetadata can come from any or all of 3 placesInput Modules -MetadataInput Modules -Metadata1) A regular expression based on the file or folder name (see module help for resources on how to configure) Input Modules -Metadata2) Extracted from the file headers (make sure to hit the “Extract Metadata” button!)Input Modules -Metadata3) Brought in from an internal file-can’t be used on its own, can only be used in addition to 1 or 2Input Modules -Metadata\n",
      "When you’re done, hit “Update”, and the metadata will populate in the boxInput Modules -NamesAndTypesSet if you have one vs >1 images per image set Set 2D vs 3DDetermine how to recognize each channel (iemetadata, file name, etc)Give each channel a descriptive nameSet the image type (Grayscale, color, object label matrix, etc)Input Modules -GroupsKeep this off if not usingYou need it if it matters which images are processed togetherInput Modules -GroupsSet variable(s) to group byImage Processing Modules●All classical image analysis segmentation works off ofthe principle-“my image should have my objects of interest bright and everywhere else dark”●For segmentation purposes, basically any transformation you can think of to achieve this is “legal”○Emphatically, this does not apply to images you are actively measuring, just images to be fed into a segmentation●To achieve this, you may need to enhance dim signal that really is there and/or mask out bright signal that truly is (but is not supposed to be segmented at the current time)●The more transformations you do before segmentation, the more critical it is to check your segmentations against your raw images (with OverlayOutlines, OverlayObjects, or the WorkspaceViewer)Overlaying segmentations on raw images is critical (but easy!)\n",
      "Image Processing modules in CellProfiler●In the Image Processing category:○EnhanceEdges○EnhanceOrSuppressFeatures-works on tubes, speckles, and/or dark holes○Smooth○Crop○ColorToGray/ GrayToColor●In the “Advanced” catgory:○GaussianFilter○MedialAxis○RemoveHoles○GaussianFilter○MatchTemplate○RunImageJMacro(Not exhaustive, just inspirational)Primary object identification\n",
      "Many options for thresholding, splitting and merging, etc.Primary object identification\n",
      "Which image to use, and what to call the objectsPrimary object identificationUse defaults for thresholding, splitting and merging, etc.\n",
      "Primary object identificationYou have options as to which objects to keep, how to threshold, etc\n",
      "Primary object identificationOutline colors show if objects are rejected (and why)\n",
      "Green= accepted (shows in the top right)Pink= not accepted (outside of the size range (too large OR too small), does not show in the top right)Orange = not accepted (touching the edge, does not show in the top right)Object identification●Once the images are loaded, how do you find objects of interest?•Step 1:Distinguish the foreground from the background (thresholding)•Step 2:Split/merge “objects” properly\n",
      "Primary object identification\n",
      "How should we threshold the image?Pixel valuesFrequencyImage thresholding●Definition:Division of the image into background and foregroundMethods:–Minimum Cross Entropy: Good for readily identifiable foreground / background –Otsu: Choose between 2-or 3-class if mid-level intensities present–Robust Background: Good for images in which most of the image is comprised of background\n",
      "What is the best threshold value for dividing the intensity into foreground and background pixels?\n",
      "Primary object identification\n",
      "What adjustments and limits should be applied to the threshold?Image thresholdingCorrection factors can nudge a threshold higher or lowerWhat is the best threshold value for dividing the intensity into foreground and background pixels?Pixel valuesFrequencyCorrection = 0.8CalculatedCorrection = 1.1Image thresholdingUpper and lower bounds can help with odd histogramsWhat is the best threshold value for dividing the intensity into foreground and background pixels?Pixel valuesFrequencyLower boundUpper boundImage thresholdingUpper and lower bounds can help with odd histograms\n",
      "Pixel valuesFrequencyLower boundUpper boundCurrent thresholdPositive pixels\n",
      "Image thresholding\n",
      "Pixel valuesFrequencyLower boundUpper boundCurrent thresholdPositive pixelsUpper and lower bounds can help with odd histogramsImage thresholding\n",
      "Pixel valuesFrequencyLower boundUpper boundCurrent thresholdPositive pixels\n",
      "Upper and lower bounds can help with odd histogramsImage thresholding\n",
      "Pixel valuesFrequencyLower boundUpper boundCurrent thresholdPositive pixels\n",
      "Upper and lower bounds can help with odd histogramsPrimary object identification\n",
      "Should we transform the image before thresholding?Image thresholdingTransformations to the image can make it easier to threshold\n",
      "“Log transform before thresholding?” selected?NoYes Separating touching objects\n",
      "●We next need to distinguish multiple objects contained in the same “clump”\n",
      "Images from Carolina Wahlby\n",
      "••••••••1\n",
      "••••••••Separating touching objects –clump identification\n",
      "○Intensity:Works best if objects are brighter at center, dimmer at edges\n",
      "○Shape:Works best if objects have indentations where clumps touch (esp. if objects are round)\n",
      "Peaks2\n",
      "12Indentations1Separating touching objects\n",
      "○Shape:Draws boundary lines at the indentations\n",
      "○Intensity:Draws boundary lines at dimmest line between objects\n",
      "•Drawing boundaries: Two options\n",
      "1\n",
      "●FIJI, Coloc2-https://imagej.net/Coloc_2FIJIDAPIGFP\n",
      "Objects can be measured for their… colocalization\n",
      "●FIJI, Coloc2-https://imagej.net/Coloc_2\n",
      "DAPIGFP\n",
      "Information is hidden in images\n",
      "G1G2SProphaseTelophaseAnaphaseMetaphase\n",
      "actual phase:G1SG2ProphaseMetaphaseAnaphaseTelophasepredicted phase\n",
      "Minh DoanHolger Hennig, Rostock UPaul Rees, Swansea UClaire McQuin\n",
      "Fabian Theis,    Helmholtz Zentrum, München\n",
      "Andrew Filby, Newcastle UHuw Summers, Swansea UWhat if we could use the information that we can’t see?Objects can be measured for their… Zernike Polynomials\n",
      "https://en.wikipedia.org/wiki/File:Zernike_polynomials2.png\n",
      "Zernike PolynomialsZernike polynomials can give accurate descriptions of  an object’s shape\n",
      "Boland, M.V., Markey, M.K., Murphy, R.F.1998\n",
      "Objects can be measured for their… texture\n",
      "Saito A, NumataY, Hamada T, et al. J PatholInform. 2016;7:36\n",
      "File utilities in CellProfiler■SaveImages□A variety of file types, bit depths, compression options■ExportToSpreadsheet□Saves as CSV or TSV□Can customize which measurements to save■ExportToDatabase□Saves to SQLite or MySQL□Can optionally make a properties file for CellProfilerAnalyst (more later)Running on large image sets on CellProfiler■A few –a few hundred images□Can likely run on your local machine□CellProfilerwill automatically multithread process up to your number of CPUs■A few hundred –a few tens of thousands of images□Talk to your local sysadmin about running on a cluster (directly or with Docker)□Check out our instructions on getting started■A few tens of thousands –a few million images□Can consider cloud processing□Check out our Distributed-CellProfilerpackagefor running on AWS“The thing I want to do doesn’t exist in CellProfiler!”■Are you sure it doesn’t?□Search the help, and/or post on image.sc(more later)■Is it an image processing utility that exists in ImageJ/Fiji?□Try out the RunImageJMacromodule, which can point to your system ImageJ/Fiji■Go ahead and write your own!□We have templates to expandand a video on how to do itMorphological Profiling\n",
      "Morphological profiling can predict small molecule activity\n",
      "Vebjorn LjosaRob ter Horst\n",
      "Actual classClass predicted by morphology\n",
      "AstraZeneca\n",
      "Ljosa, et al. J BiomolScreening 2013Treat cells with compoundsImagingImage analysisMatchmorphological profiles\n",
      "Morphological profiling can identify drugs for disease\n",
      "Gibson, et al. Circulation 2015; Anne Carpenter is a board member at Recursion Pharmaceuticals\n",
      "Healthy \n",
      "Disease (CCM knockdown)+ drug?\n",
      "Disease states can be morphologically distinct\n",
      "Donna McPhie\n",
      "Bruce Cohen, McLean HospitalRakesh Karma-charyaMohammad RohbanKyle Karhohs\n",
      "MarziehHaghighi\n",
      "Controls\n",
      "Bipolar\n",
      "Depression\n",
      "Schizophrenia\n",
      "Schizoaffective\n",
      "Disease states can be morphologically distinctEven in subtle ways not obvious to the naked eye\n",
      "Disease states can be morphologically distinct\n",
      "Low ScorersHigh ScorersEven in subtle ways not obvious to the naked eyeCellProfiler-Analyst: overviewExploredata large sets of imagesIdentifyinteresting subpopulations and see the original imagesIdentify interesting phenotypes automatically\n",
      "Goal:Provide the user with a powerful suite of image exploration and machine learning methods\n",
      "How can I learn how to do this stuff?Where can I go for help?forum.image.sc-Open scientific community forum for bioimage analysis and beyond\n",
      "Center for Open Bioimage AnalysisOpenbioimageanalysis.org\n",
      "Gratitude\n",
      "Recent major funding for this work provided by:●CZI Imaging Scientist Fellowship●NIH NIGMS: MIRA R35 GM122547●CZI Software Fellows program●NIH NIGMS: P41 GM135019Many thanks to our many biology collaboratorsBeth CiminiMario Costa CruzBarbara Diaz-RohrerTeresa GaoMelissaGillisNodarGogoberidzeSerena LarewBecki LedfordLevin MoserRebecca SenftCallumTromans-CoiaErin WeisbartAnne CarpenterShantanu SinghJohn ArevaloNiranjChandrasekaranMarzieh HaghighiYu HanAlexander KalininSerena LarewBecki LedfordRobert van DijkCimini Lab members\n",
      "IMAGINGPLATFORM\n",
      "Carpenter-Singh Lab members\n",
      "\n",
      "••••••••Primary object identificationAdjusting the sizes of our expected objects\n",
      "Separating touching objects\n",
      "●Smoothing: Increase to reduce intensity irregularities which produce over-segmentation of objects \n",
      "Original imageSmoothing filtersize = 4Smoothing filtersize = 8Separating touching objects\n",
      "●Suppress Local Maxima\n",
      "●Smallest distance allowed between object intensity peaks to be considered one object rather than a clump●Decrease to reduce improper merging of objects in clumps\n",
      "Original imageMaxima distance = 4Maxima distance = 8MaximaSeparating touching objects\n",
      "●Over-adjusting can produce more improper segmentation than it solves●The proper settings are usually a matter of trial and error, and trying to globally maximize good segmentations rather than tuning to one imageHowever….\n",
      "Primary object identificationMaxima can be displayed to assist in creating objects\n",
      "Primary object identification\n",
      "Primary object identificationWhen if at allshould you fill holes in the objects?\n",
      "Primary object identification\n",
      "Hole filling “After declumpingonly”Hole filling “After both thresholding and declumping”When if at allshould you fill holes in the objects?Secondary object identification●Make a second object based on a first oneOur colors mean different things here!Green= the input primary objectPink= the secondary object\n",
      "Many options for thresholding, splitting and merging, etc.Secondary object identification●A few ways to do it●Non-image-based:○Distance-N –draw a circle a fixed width around the primary object (1)●Image-based-all start by thresholding the image, then:○Distance-B –draw a circled a fixed with around the primary object, but only into areas that are threshold positive (2)○Watershed-Image and Watershed-Gradient –fill the thresholdedarea in a watershed, either based on the original pixel data or a transform of it (3)○Propagation-do watershed image, but balance those lines against trying to make each cell the same size (4)\n",
      "1\n",
      "2\n",
      "3\n",
      "4Tertiary object identificationMake a third object out of two othersThe third object should be the differences of the first twoThe second object should therefore also be derived from the first object (such as by IdentifySecondaryObjectsor ExpandOrShrinkObjects)\n",
      "Object processing modules in CellProfiler●In the Object Processing category:○ExpandOrShrinkObjects○FilterObjects○MaskObjects○RelateObjects○EditObjectsManually○CombineObjects●In the “Advanced” catgory:○DilateObjects○ErodeObjects○ShrinkToObjectCenters(Not exhaustive, just inspirational)How can we push this further?Are there ways we can better find objects?Stardist\n",
      "Martin Weigert, Uwe Schmidt, Robert Haase, Ko Sugawara, and Gene MyersCellpose\n",
      "Carsen Stringer and Marius Pachitariu\n",
      "Cellpose2\n",
      "Carsen Stringer and Marius Pachitariu\n",
      "Future models may not need so much ground truth -SPOCO\n",
      "Adrian Wolny, Qin Yu, Constantin Pape, Anna Kreshuk\n",
      "Tips for creating a good high content analysis workflow●When finding the objects that you care about, ask yourself for your whole experiment:■Do I generally agree with most of the object segmentations from my analysis workflow? ■Do I have an approximately equal number of regions/images where the threshold chosen by the algorithm for this imageis a bit too low vs a bit too high?■Do I have an approximately equal number of oversegmentations/splits and undersegmentations/merges?■Very important:Do both the second and third bullet points hold true for both my negative control images and my positive control (or most extreme expected phenotype(s) sample) images?https://carpenter-singh-lab.broadinstitute.org/blog/when-to-say-good-enoughWe’ve found our objects-now what?Feature extraction●How do you get hundreds of measurements per cell?\n",
      "Counts, Shapes, Sizes, Intensities, Textures, Correlations, RelationshipsSoftware can help us build up big measurement suites for our images\n",
      "Many of these metrics can be found in LOTS of toolsObjects can be measured for their… intensity●Integrated intensity:  Sum of the pixel intensities within an object●Mean, median, standard deviation intensities●Maximal and minimal pixel intensities●Lower/Upper quartile of the intensity●Object intensities may be measured from any channel, not just the channel used to identify the object○Example: GFP intensity may be measured using nuclei objects identified with DAPIObjects can be measured for their… intensity distribution\n",
      "Objects can be measured for their… size and shape●Area●Perimeter●Eccentricity (circle = 1, line = 0)●MajorAxisLength●MinorAxisLength●Orientation●Solidity\n",
      "Objects can be measured for their… colocalization\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mimage_analyst_agent\u001B[0m (to image_analyst_proxy_agent):\n",
      "\n",
      "As an image analyst, I analyze and interpret data from images, specifically related to microscopic, radiographic, and other visual data. Here are the tasks I could assist you with:\n",
      "\n",
      "1. **Image Segmentation**: To separate an image into regions or categories, which may correspond to different objects or parts of objects. It's a key step in image analysis intending to partition an image into meaningful parts.\n",
      "\n",
      "2. **Image Enhancement**: Modification of an image to make it more suitable for a specific application. Techniques could include brightness and contrast adjustments, noise removal, and sharpening.\n",
      "\n",
      "3. **Feature Extraction**: Identifying and extracting important characteristics from an image. These features can represent shape, texture, color, or other aspects of the image.\n",
      "\n",
      "4. **Image Registration**: The process of transforming different images or datasets into one coordinate system. It's used when we want to integrate data collected at different times, from different perspectives, or on different sensors.\n",
      "\n",
      "5. **Measuring Objects**: Assessment of specific parameters related to the objects of interest in an image, such as size, shape, color, and spatial relationships. \n",
      "\n",
      "6. **Image Classification**: Using machine learning techniques to automatically identify the class or attribute of regions or objects within an image.\n",
      "\n",
      "7. **Annotating images**: Marking images with labels or notes so that machine learning algorithms can process the image.\n",
      "\n",
      "My role plays a crucial part in fields like healthcare, remote sensing, artificial intelligence, and much more where image data is imperative.\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# 1. create an RetrieveAssistantAgent instance named \"assistant\"\n",
    "image_analyst_assistant = RetrieveAssistantAgent(\n",
    "    name=\"image_analyst_agent\", \n",
    "    system_message=\"Hi there, You are the image analyst! Your specialty is processing and analyzing microscopic, radiographic, and other visual data. You can help with tasks like image segmentation, feature extraction, image registration, measuring objects in images, and classifying/annotating images using machine learning. Please provide assistance on analyzing, enhancing, or extracting meaningful information from my imaging data!\",\n",
    "    llm_config={\n",
    "        \"timeout\": 600,\n",
    "        \"config_list\": config_list,\n",
    "    },\n",
    ")\n",
    "\n",
    "image_analyst_proxyagent = RetrieveUserProxyAgent(\n",
    "    name=\"image_analyst_proxy_agent\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    max_consecutive_auto_reply=10,\n",
    "    retrieve_config={\n",
    "        \"docs_path\": \"ImageAnalystPDF\",  # change this to your own path, such as https://raw.githubusercontent.com/microsoft/autogen/main/README.md\n",
    "        \"chunk_token_size\": 1000,\n",
    "        \"model\": config_list[0][\"model\"],\n",
    "        \"client\": chromadb.PersistentClient(path=\"chroma_image_analyst_db\"),\n",
    "        \"embedding_model\": \"all-mpnet-base-v2\",\n",
    "        \"get_or_create\": True,  # set to False if you don't want to reuse an existing collection, but you'll need to remove the collection manually\n",
    "    },\n",
    ")\n",
    "\n",
    "query = \"explain your role.\"\n",
    "\n",
    "image_analyst_proxyagent.initiate_chat(image_analyst_assistant, problem=query)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T09:38:27.247145Z",
     "start_time": "2023-11-12T09:38:04.622016Z"
    }
   },
   "id": "28bf2b0c797293d3"
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trying to create collection.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "max_tokens is too small to fit a single line of text. Breaking this line:\n",
      "\tNature Reviews Methods Primers |             (2022) 2:94 1 ...\n",
      "Failed to split docs with must_break_at_empty_line being True, set to False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doc_ids:  [['doc_21', 'doc_36', 'doc_20', 'doc_9', 'doc_39', 'doc_11', 'doc_38', 'doc_12', 'doc_13', 'doc_37', 'doc_28', 'doc_26', 'doc_40', 'doc_0', 'doc_35', 'doc_41', 'doc_10', 'doc_18', 'doc_22', 'doc_8']]\n",
      "\u001B[32mAdding doc_id doc_21 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_36 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_20 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_9 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_39 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_11 to context.\u001B[0m\n",
      "\u001B[32mAdding doc_id doc_38 to context.\u001B[0m\n",
      "\u001B[33mresearch_assistant_proxy_agent\u001B[0m (to research_assistant_agent):\n",
      "\n",
      "You're a retrieve augmented chatbot. You answer user's questions based on your own knowledge and the\n",
      "context provided by the user. You should follow the following steps to answer a question:\n",
      "Step 1, you estimate the user's intent based on the question and context. The intent can be a code generation task or\n",
      "a question answering task.\n",
      "Step 2, you reply based on the intent.\n",
      "If you can't answer the question with or without the current context, you should reply exactly `UPDATE CONTEXT`.\n",
      "If user's intent is code generation, you must obey the following rules:\n",
      "Rule 1. You MUST NOT install any packages because all the packages needed are already installed.\n",
      "Rule 2. You must follow the formats below to write your code:\n",
      "```language\n",
      "# your code\n",
      "```\n",
      "\n",
      "If user's intent is question answering, you must give as short an answer as possible.\n",
      "\n",
      "User's question is: explain your role.\n",
      "\n",
      "Context is: replicates to get a significance between conditions219 ,220. Data should \n",
      "be reported correctly as a mean of each replicate, with each replicate \n",
      "being an average of a similar number of organoids. Recently, several \n",
      "publications have provided guidelines to improve reproducibility, \n",
      "including general biological reporting and accurate statistics221 –223. \n",
      "SuperPlots is recommended for data visualization, as it shows both the  \n",
      "mean of each independent biological replicate and the spread of  \n",
      "the individual data points224 ,225. This type of plot is ideal for organoid \n",
      "reporting, as it allows immediate visualization of any potential arte -\n",
      "fact concerning batch variation. Journal collections that deepen all \n",
      "aspects of statistics used in biological sciences are valuable resources \n",
      "to facilitate accurate reporting226.\n",
      "Organoids should be characterized in full, not only checked for \n",
      "the expression of markers but also benchmarked to the tissue of ori -\n",
      "gin. It is crucial to not only determine whether cells have similar mor-\n",
      "phology and expression patterns as the tissue of origin and that they \n",
      "have a similar function (using, for example, ELISA for albumin, bile \n",
      "acid production by liver/hepatocyte organoids). Whenever possible, \n",
      "results between compared conditions should be normalized against \n",
      "DNA content (for genomic analysis) or structural proteins (such as \n",
      "actin, for western blotting), whereas functional data such as from ELISA \n",
      "should be normalized to the total number of cells or area of organoid \n",
      "structures in each well being compared. Improving reporting standards \n",
      "is the first step towards increasing reproducibility, thus enabling the  \n",
      "field to move forward.Data deposition\n",
      "Standardized organoid data repositories are currently missing. A nota -\n",
      "ble exception is a patient-derived repository for cancer organoids led \n",
      "by the National Cancer Institute (NCI) in the USA and the Human Cancer \n",
      "Models Initiative (HCMI), a combined international effort of several \n",
      "institutions (Cancer Research UK; NCI; Hubrecht Organoid Technol -\n",
      "ogy’s the HUB; and the US Office of Cancer Genomics’ OCG) (Supple-\n",
      "mentary Table 3). However, healthy patient-derived organoids and \n",
      "organoid models from non-cancer diseases, both human and murine, \n",
      "are yet to be documented in one database, indicating the apparent \n",
      "need to fully document, report and deposit all organoid lines gener -\n",
      "ated. The recent launch of the Human Cell Atlas-Organoid initiative, \n",
      "documenting all human-derived organoids227, is the first step in that \n",
      "direction. For transcriptomics-related data, such as gene expression, \n",
      "non-coding RNA, ChIP, genome methylation, high-throughput PCR \n",
      "with reverse transcription, SNP arrays, SAGE and protein arrays, the \n",
      "Gene Expression Omnibus (GEO) is generally recommended228 ,229. For \n",
      "code used to analyse organoid data sets, GitHub is recommended. We \n",
      "envision that data and organoid depositories will arise shortly to fill in \n",
      "the presently missing gap in data repositories for organoid work. When \n",
      "depositing data from human organoids, it is important to consider \n",
      "all ethical regulations and the necessity of obtaining patient consent \n",
      "and keeping patient anonymity. Critical points have been extensively \n",
      "138. Raudvere, U. et al. g:Profiler: a web server for functional enrichment analysis and \n",
      "conversions of gene lists (2019 update). Nucleic Acids Res. 47, W191–W198 (2019).\n",
      "139. Powers, R. K., Goodspeed, A., Pielke-Lombardo, H., Tan, A. C. & Costello, J. C.  \n",
      "GSEA-InContext: identifying novel and common patterns in expression experiments. Bioinformatics 34, i555–i564 (2018).\n",
      "140. Shannon, P. et al. Cytoscape: a software environment for integrated models of \n",
      "biomolecular interaction networks. Genome Res. 13, 2498–2504 (2003).\n",
      "141. Reimand, J. et al. Pathway enrichment analysis and visualization of omics data using g:Profiler, GSEA, Cytoscape and EnrichmentMap. Nat. Protoc. 14, 482–517 (2019).\n",
      "142. Berg, S. et al. ilastik: interactive machine learning for (bio)image analysis. Nat. Methods  \n",
      "16, 1226–1232 (2019).\n",
      "143. Kok, R. N. U. et al. OrganoidTracker: efficient cell tracking using machine learning and manual error correction. PLoS ONE 15, e0240802 (2020).\n",
      "144. Mergenthaler, P. et al. Rapid 3D phenotypic analysis of neurons and organoids using data-driven cell segmentation-free machine learning. PLoS Comput. Biol. 17, e1008630 \n",
      "(2021).\n",
      "145. Loomans, C. J. M. et al. Expansion of adult human pancreatic tissue yields organoids harboring progenitor cells with endocrine differentiation potential. Stem Cell Rep. 10, \n",
      "712–724 (2018).\n",
      "146. Elizondo, D. M. et al. Pancreatic islets seeded in a novel bioscaffold forms an organoid  \n",
      "to rescue insulin production and reverse hyperglycemia in models of type 1 diabetes.  \n",
      "Sci. Rep. 10, 4362 (2020).\n",
      "147. Meran, L. et al. Engineering transplantable jejunal mucosal grafts using patient-derived organoids from children with intestinal failure. Nat. Med. 26, 1593–1601 (2020).148. Sugimoto, S. et al. An organoid-based organ-repurposing approach to treat short bowel syndrome. Nature 592, 99–104 (2021).\n",
      "149. Vives, J. & Batlle-Morera, L. The challenge of developing human 3D organoids into medicines. Stem Cell Res. Ther. 11, 72 (2020).\n",
      "150. Phan, N. et al. A simple high-throughput approach identifies actionable drug sensitivities \n",
      "in patient-derived tumor organoids. Commun. Biol. 2, 78 (2019).\n",
      "151. Tebon, P. J. et al. Drug screening at single-organoid resolution via bioprinting and \n",
      "interferometry. Preprint at bioRxiv  https://doi.org/10.1101/2021.10.03.462896 (2021).\n",
      "152. Chen, J. et al. An organoid-based drug screening identified a menin-MLL inhibitor for endometrial cancer through regulating the HIF pathway. Cancer Gene Ther. 28, 112–125 \n",
      "(2021).\n",
      "153. Verissimo, C. S. et al. Targeting mutant RAS in patient-derived colorectal cancer organoids by combinatorial drug screening. eLife https://doi.org/10.7554/eLife.18489 \n",
      "(2016).\n",
      "154. Nuciforo, S. et al. Organoid models of human liver cancers derived from tumor needle biopsies. Cell Rep. 24, 1363–1376 (2018).\n",
      "155. Schnalzger, T. E. et al. 3D model for CAR-mediated cytotoxicity using patient-derived \n",
      "colorectal cancer organoids. EMBO J. 38, 100928 (2019).\n",
      "156. Neal, J. T. et al. Organoid modeling of the tumor immune microenvironment. Cell 175, \n",
      "1972–1988.e16 (2018).\n",
      "157. Weeber, F. et al. Preserved genetic diversity in organoids cultured from biopsies  \n",
      "of human colorectal cancer metastases. Proc. Natl Acad. Sci. USA 112, 13308–13311 \n",
      "(2015).\n",
      "(reviewed elsewhere211, 212) — such as the recently published endothelial \n",
      "and healthy human colon and human colorectal cancer organoids213, \n",
      "murine cholangiocyte and mesenchymal organoids73 and murine \n",
      "pancreatic islet and endothelial organoids109 — will require controlling \n",
      "sources of variation derived from incorporating additional cells to an \n",
      "already heterogeneous system. Exerting control over the variation \n",
      "while generating more functionally relevant structures represents a \n",
      "significant challenge for the field.\n",
      "Finally, significant heterogeneity is also observed following orga -\n",
      "noid treatments. Although this represents a drawback when trying to \n",
      "use organoids as a tool for investigating molecular mechanism, it is \n",
      "beneficial when aiming to model patient responses. In particular, for cancer treatment, several different patient-derived organoid models \n",
      "tested (including, but not limited to, pancreas, colorectal and liver \n",
      "cancer) exhibit heterogeneous responses to known chemotherapeutic Nature Reviews Methods Primers |             (2022) 2:94 13\n",
      "Primerand anticancer agents48,131, 214. This parallels the variability in patient \n",
      "outcomes observed in the clinic and has positioned cancer organoids \n",
      "as a potential predictive tool for drug testing (reviewed elsewhere215). \n",
      "This heterogeneity in treatment responses is now being exploited for \n",
      "the treatment of patients with cystic fibrosis193, with a large ongoing \n",
      "European multicentre clinical trial study, led by the HIT-Cystic Fibrosis \n",
      "consortium, to identify potential responders216. Similar initiatives are \n",
      "lagging behind for patients with cancer, although mostly because of \n",
      "disparities in establishment rates and lengthy timelines (reviewed \n",
      "elsewhere217). Organoids were recently used to inform the treatment \n",
      "of a patient suffering from a rare form of liver cancer, although unfor-\n",
      "tunately the treatment was discontinued due to deterioration of the \n",
      "patient’s general condition, preventing the authors from drawing \n",
      "conclusions regarding the predictive value of the organoid system \n",
      "for drug treatment218.\n",
      "Minimum reporting standards\n",
      "To improve the reproducibility and reliability of the results, it is essen -\n",
      "tial to account for the variability between different cell isolations and \n",
      "intra-organoid heterogeneity. Although one could consider that at \n",
      "least three independent experiments with at least three different bio -\n",
      "logical replicates are a minimum requirement, this may not always \n",
      "be sufficient. In fact, the number of organoids to study per biological \n",
      "replicate as well as the number of biological replicates depends on  \n",
      "the research question, the specific experiment and the variability  \n",
      "of the phenotype observed. Whenever possible, calculating the num-\n",
      "ber of independent organoids per biological replicate, sample size \n",
      "calculations similar to those used for mouse experiments, taking into \n",
      "account statistical power and variance of the experiment, would be \n",
      "highly recommended. The data should be robust, meaning repeated \n",
      "in at least three independent biological replicates, which refers to \n",
      "independent isolations from different animals or patients. However, \n",
      "replicates and the research question for each experiment should be \n",
      "determined in advance to avoid hypothesis fishing and increasing \n",
      "structure should be present within the organoids, and intra-cellular \n",
      "hormone vesicles109 are also expected to be observed. For branching \n",
      "epithelial organoids, such as breast organoids110, distinct branching \n",
      "structures are expected. For intestinal organoids, crypt-like struc -\n",
      "tures99 should be observed. For more complex organoid culture sys -\n",
      "tems comprising incorporated supportive cells, such as fibroblast111 \n",
      "and/or endothelial cells (ECs)106 ,109, it is expected that the incorpo -\n",
      "rated ECs should form a vascular network106 that recapitulates the  \n",
      "endothelial–epithelial interactions around the organoids.\n",
      "Organoid functionality\n",
      "Evaluation of organoid functionality includes, but is not limited to, \n",
      "the generation of mature cells, the formation of the vasculature or \n",
      "neuronal networks25,43, the accurate response to external stimuli and Nature Reviews Methods Primers |             (2022) 2:94 7\n",
      "Primerthe effective secretion of cytokines or hormones, among other fac -\n",
      "tors. Organoids should be compared with freshly isolated tissues as a \n",
      "positive control. Discrete physiological recapitulations are required for \n",
      "different tissues, such as acid secretion in gastric organoids and mucus \n",
      "secretion in intestinal organoids. Various methods of characterization \n",
      "to assess the structure and functions of organoids are listed in Table  1.\n",
      "Pancreatic islet organoids. We use the example of mouse pancreatic \n",
      "islet organoids to discuss various characterization and validation \n",
      "methods (Fig.  3). The formation of four endocrine differentiated cell \n",
      "types (β-cell, α-cell, δ-cell and PP cell) is examined by the expressions of the corresponding hormones — insulin, glucagon, somatostatin \n",
      "and pancreatic polypeptide, respectively — by immunostaining112, 113. \n",
      "To assess the extent of β-cell maturation, the ultrastructural mor -\n",
      "phometric analysis of insulin secretory granules is performed by \n",
      "immuno-gold transmission electron microscopy (TEM) imaging109. \n",
      "To evaluate the ability of the β-cells in organoid responsiveness to \n",
      "glucose, glucose-stimulated insulin secretion experiments are per -\n",
      "formed, for example using an insulin or a C-peptide ELISA assay with \n",
      "cyclic high and low glucose incubation114. As insulin release is asso -\n",
      "ciated with calcium dynamics, calcium signalling traces imaging \n",
      "rapidly increases in response to glucose and returns to baseline113 ,115.  \n",
      "Table 1 | Methods used in organoid research to assess/characterize organoid structure and function\n",
      "Result Characterization methods Function Example applications Refs.\n",
      "Organoid structure Bright-field imaging Qualitative assessment of morphology \n",
      "and viability; quantitative assessment of organ size and numberCrypt-like structures for intestinal organoids, islet-like globular structure for pancreatic islet organoids, branching structure for breast or \n",
      "lung organoids, hollow cystic structures for \n",
      "cholangiocyte and CC organoids, grape-like \n",
      "structures for hepatocyte and HCC organoids34–36,41,99, \n",
      "109,110,124, \n",
      "125,131,154, \n",
      "203,205\n",
      "Immunofluorescent staining Quantitation of viable cells; Edu and \n",
      "Ki67 staining to evaluate the survival \n",
      "and proliferation of the cells in the \n",
      "organoids; staining of differentially \n",
      "expressed proteins in a different layer or \n",
      "geometry; qualitative and quantitative \n",
      "assessment of cellular composition, \n",
      "spatial distribution and proportion of different cell types and maturation state \n",
      "of the different subpopulationsBile canaliculi can be shown to form between \n",
      "the hepatocytes (identified by markers such \n",
      "as BSEP or MRP2 for bile canaliculi, ALB or \n",
      "197. Howell, K. et al. DNA methylation and transcription patterns in intestinal epithelial  \n",
      "cells from pediatric patients with inflammatory bowel diseases differentiate disease  \n",
      "subtypes and associate with outcome. Gastroenterology https://doi.org/10.1053/  \n",
      "j.gastro.2017.10.007 (2017).\n",
      "198. Soroka, C. et al. Bile‐ derived organoids from patients with primary sclerosing cholangitis \n",
      "recapitulate their inflammatory immune profile. Hepatology https://doi.org/10.1002/\n",
      "hep.30470 (2018).\n",
      "199. Danahay, H. et al. Notch2 is required for inflammatory cytokine-driven goblet cell \n",
      "metaplasia in the lung. Cell Rep. 10, 239–252 (2015).\n",
      "200. Seidlitz, T. et al. Mouse models of human gastric cancer subtypes with stomach-specific \n",
      "CreERT2-mediated pathway alterations. Gastroenterology 157, 1599–1614.e2 (2019).\n",
      "201. Sachs, N. et al. A living biobank of breast cancer organoids captures disease \n",
      "heterogeneity. Cell 172, 373–386.e10 (2018).\n",
      "202. Boj, S. F. et al. Organoid models of human and mouse ductal pancreatic cancer. Cell 160, \n",
      "324–338 (2015).\n",
      "203. Rimland, C. A. et al. Regional differences in human biliary tissues and corresponding in \n",
      "vitro-derived organoids. Hepatology 73, 247–267 (2021).\n",
      "204. Sampaziotis, F. et al. Reconstruction of the mouse extrahepatic biliary tree using primary \n",
      "human extrahepatic cholangiocyte organoids. Nat. Med. 23, 954–963 (2017).\n",
      "205. Sampaziotis, F. et al. Cholangiocyte organoids can repair bile ducts after transplantation in the human liver. Science 371, 839–846 (2021).\n",
      "206. Serra, D. et al. Self-organization and symmetry breaking in intestinal organoid \n",
      "development. Nature 569, 66–72 (2019).\n",
      "207. Kozlowski, M. T., Crook, C. J. & Ku, H. T. Towards organoid culture without Matrigel. \n",
      "Commun. Biol. 4, 1387 (2021).\n",
      "208. Aloia, L. et al. Epigenetic remodelling licences adult cholangiocytes for organoid formation and liver regeneration. Nat. Cell Biol. 21, 1321–1333 (2019).\n",
      "209. Wang, Y. et al. A microengineered collagen scaffold for generating a polarized  \n",
      "crypt–villus architecture of human small intestinal epithelium. Biomaterials 128,  \n",
      "44–55 (2017).\n",
      "210. Lukonin, I. et al. Phenotypic landscape of intestinal organoid regeneration. Nature 586, \n",
      "275–280 (2020).\n",
      "211. Kretzschmar, K. & Clevers, H. Organoids: modeling development and the stem cell niche in a dish. Dev. Cell 38, 590–600 (2016).\n",
      "212. Kim, J., Koo, B. K. & Knoblich, J. A. Human organoids: model systems for human biology and medicine. Nat. Rev. Mol. Cell Biol. 21, 571–584 (2020).\n",
      "213. Palikuqi, B. et al. Adaptable haemodynamic endothelial cells for organogenesis and \n",
      "tumorigenesis. Nature 585, 426–432 (2020).\n",
      "214. Tiriac, H. et al. Organoid profiling identifies common responders to chemotherapy in \n",
      "pancreatic cancerpancreatic cancer organoids parallel patient response. Cancer Discov.  \n",
      "8, 1112–1129 (2018).\n",
      "215. Wensink, G. E. et al. Patient-derived organoids as a predictive biomarker for treatment response in cancer patients. NPJ Precis. Oncol. 5, 30 (2021).\n",
      "216. van Mourik, P. et al. Rationale and design of the HIT-CF organoid study: stratifying cystic fibrosis patients based on intestinal organoid response to different CFTR-modulators. Transl. Med. Commun. https://doi.org/10.1186/s41231-020-00060-3 (2020).\n",
      "of normalized and stable blood glucose level, normal plasma insulin level and maintained \n",
      "body weight34,40,109, \n",
      "113,115,116, \n",
      "124,125,131, \n",
      "154,205\n",
      "CC, cholangiocarcinoma; HCC, hepatocellular carcinoma; scRNA-seq, single-cell RNA sequencing.Nature Reviews Methods Primers |             (2022) 2:94 8\n",
      "PrimerTo test the full potential of the islet organoids upon transplantation, \n",
      "in vivo functional evaluation is needed to ameliorate the hypergly -\n",
      "caemic phenotypes of  streptozotocin-induced type 1 diabetic mouse models109 ,113 ,116. The characterization parameters include normal -\n",
      "ized and stable blood glucose level, normal plasma insulin level and  \n",
      "maintained body weight.\n",
      "25 /uni03BCm\n",
      " 100 /uni03BCm 100 /uni03BCmECs\n",
      "ECs\n",
      "25 /uni03BCm 25 /uni03BCm ECsECs\n",
      "Organoid\n",
      "OrganoidOrganoid\n",
      "Organoids RNA\n",
      "Diabetic mouse model2 mM glucose 20 mM glucose Wash WashThree rounds\n",
      "A  Day 7 (organoids growing in matrigel)\n",
      "Aa\n",
      "CaAb\n",
      "Cb CcAc\n",
      "C  Day 30 (diﬀerentiated organoids with ECs)\n",
      "E F G\n",
      "HD  B  \n",
      "Ins DAPI Gcg DAPI Sst DAPI Ppy DAPI\n",
      "2 20 2 20 2 205\n",
      "01015\n",
      "Sample 1 Sample 2\n",
      "Quantitative PCRRelative to HPRTGene\n",
      "***\n",
      "5\n",
      "01015\n",
      "Three rounds of glucose\n",
      "stimulation (mM)C-peptide release\n",
      "(ng per 105 cells)**** ***\n",
      "Function validation in vitro\n",
      "• C-peptide• Calcium image\n",
      "Function validation in vivo• Blood glucose• Glucose tolerance test• Plasma insulin level• Body weight0300\n",
      "01020304050120\n",
      "100\n",
      "80\n",
      "60Intensity\n",
      "Time (s)2 mM glucose\n",
      "20 mM glucose\n",
      "30 mM KCl\n",
      "Fresh isletsor organoids\n",
      "Kidney\n",
      "Body\n",
      "Organoid transplantation\n",
      "Body\n",
      "177. Heo, I. et al. Modelling Cryptosporidium infection in human small intestinal and lung organoids. Nat. Microbiol. 3, 814–823 (2018).\n",
      "178. Chen, S. et al. Rotavirus infection and cytopathogenesis in human biliary organoids \n",
      "potentially recapitulate biliary atresia development. mBio https://doi.org/10.1128/\n",
      "mBio.01968-20 (2020).\n",
      "179. Sachs, N. et al. Long-term expanding human airway organoids for disease modeling. EMBO J. 38, 100300 (2019).\n",
      "180. Driehuis, E. et al. Oral mucosal organoids as a potential platform for personalized cancer \n",
      "therapy. Cancer Discov. 9, 852–871 (2019).\n",
      "181. Wroblewski, L. E. et al. Helicobacter pylori targets cancer-associated apical–junctional constituents in gastroids and gastric epithelial cells. Gut 64, 720–730 (2015).Nature Reviews Methods Primers |             (2022) 2:94 20\n",
      "Primer182. Co, J. Y. et al. Controlling epithelial polarity: a human enteroid model for host-pathogen \n",
      "interactions. Cell Rep. 26, 2509–2520.e4 (2019).\n",
      "183. Saxena, K. et al. Human intestinal enteroids: a new model to study human rotavirus infection, host restriction, and pathophysiology. J. Virol. 90, 43–56 (2016).\n",
      "184. Salahudeen, A. A. et al. Progenitor identification and SARS-CoV-2 infection in human \n",
      "distal lung organoids. Nature 588, 670–675 (2020).\n",
      "185. Lamers, M. M. et al. SARS-CoV-2 productively infects human gut enterocytes. Science  \n",
      "369, 50–54 (2020).\n",
      "186. Zang, R. et al. TMPRSS2 and TMPRSS4 promote SARS-CoV-2 infection of human small \n",
      "intestinal enterocytes. Sci. Immunol. 5, 3582 (2020).\n",
      "187. Van Goor, F. et al. Correction of the F508del-CFTR protein processing defect in vitro by \n",
      "the investigational drug VX-809. Proc. Natl Acad. Sci. USA 108, 18843–18848 (2011).\n",
      "188. Van Goor, F. et al. Rescue of CF airway epithelial cell function in vitro by a CFTR \n",
      "potentiator, VX-770. Proc. Natl Acad. Sci. USA 106, 18825–18830 (2009).\n",
      "189. Ratjen, F. et al. Cystic fibrosis. Nat. Rev. Dis. Primers 1, 15010 (2015).\n",
      "190. Dekkers, J. F. et al. A functional CFTR assay using primary cystic fibrosis intestinal organoids. Nat. Med. 19, 939–945 (2013).\n",
      "191. Verstegen, M. M. A. et al. Human extrahepatic and intrahepatic cholangiocyte organoids show region-specific differentiation potential and model cystic fibrosis-related bile duct disease. Sci. Rep. 10, 21900 (2020).\n",
      "192. Dekkers, J. et al. Characterizing responses to CFTR-modulating drugs using rectal organoids derived from subjects with cystic fibrosis. Sci. Transl. Med. 8, 344ra384 (2016).\n",
      "193. Berkers, G. et al. Rectal organoids enable personalized treatment of cystic fibrosis.  \n",
      "Cell Rep. 26, 1701–1708.e3 (2019).\n",
      "194. Brewington, J. J. et al. Detection of CFTR function and modulation in primary human nasal cell spheroids. J. Cyst. Fibros. 17, 26–33 (2018).\n",
      "195. Schwank, G. et al. Functional repair of CFTR by CRISPR/Cas9 in intestinal stem cell organoids of cystic fibrosis patients. Cell Stem Cell 13, 653–658 (2013).\n",
      "196. Boye, T., Steenholdt, C., Jensen, K. & Nielsen, O. Molecular manipulations and intestinal \n",
      "stem cell-derived organoids in inflammatory bowel disease. Stem Cells https://doi.org/  \n",
      "10.1093/stmcls/sxac014 (2022).\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mresearch_assistant_agent\u001B[0m (to research_assistant_proxy_agent):\n",
      "\n",
      "As an AI research assistant, my main role is to assist in any way to accelerate your research endeavor. I have the capability to find relevant academic papers based on given topics from a wide range of fields. I can extract key information from these sources and generate concise summaries or detailed notes on the data. Additionally, I am proficient in managing documentation (e.g., editing, formatting) and tracking project timelines to ensure deadlines are met. I can act as a laboratory assistant by helping locate required equipment, ensuring protocol adherence, and supporting lab experiments virtually. Generally, my main task is to facilitate efficiency and accuracy in your research tasks.\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# 1. create an RetrieveAssistantAgent instance named \"assistant\"\n",
    "research_assistant_assistant = RetrieveAssistantAgent(\n",
    "    name=\"research_assistant_agent\", \n",
    "    system_message=\"Greetings, You are the research assistant. You excel at literature reviews, proposal preparations, quantitative and qualitative data collection/analysis, lab experiment assistance, and research project coordination. Please help me find relevant papers, take notes, format documents, track project timelines, locate lab equipment, ensure protocol adherence, or any other tasks that will accelerate my research!\",\n",
    "    llm_config={\n",
    "        \"timeout\": 600,\n",
    "        \"config_list\": config_list,\n",
    "    },\n",
    ")\n",
    "\n",
    "research_assistant_proxyagent = RetrieveUserProxyAgent(\n",
    "    name=\"research_assistant_proxy_agent\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    max_consecutive_auto_reply=10,\n",
    "    retrieve_config={\n",
    "        \"docs_path\": \"ResearchAssistantPDF\",  # change this to your own path, such as https://raw.githubusercontent.com/microsoft/autogen/main/README.md\n",
    "        \"chunk_token_size\": 1000,\n",
    "        \"model\": config_list[0][\"model\"],\n",
    "        \"client\": chromadb.PersistentClient(path=\"chroma_research_assistant_db\"),\n",
    "        \"embedding_model\": \"all-mpnet-base-v2\",\n",
    "        \"get_or_create\": True,  # set to False if you don't want to reuse an existing collection, but you'll need to remove the collection manually\n",
    "    },\n",
    ")\n",
    "\n",
    "query = \"explain your role.\"\n",
    "\n",
    "research_assistant_proxyagent.initiate_chat(research_assistant_assistant, problem=query)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T09:42:19.619259Z",
     "start_time": "2023-11-12T09:41:56.615317Z"
    }
   },
   "id": "c046dbfb8f5875ac"
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Groupchat"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5059fa1a092dde8c"
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "models to use:  ['gpt-3.5-turbo-1106']\n",
      "vision models to use:  ['gpt-4-vision-preview']\n"
     ]
    }
   ],
   "source": [
    "import autogen\n",
    "from autogen.agentchat.contrib.retrieve_assistant_agent import RetrieveAssistantAgent\n",
    "from autogen.agentchat.contrib.retrieve_user_proxy_agent import RetrieveUserProxyAgent\n",
    "import chromadb\n",
    "\n",
    "config_list = autogen.config_list_from_json(\n",
    "    env_or_file=\"modelConfig.json\",\n",
    "    file_location=\".\",\n",
    "    filter_dict={\n",
    "        \"model\": {\n",
    "            \"gpt-3.5-turbo-1106\"\n",
    "        }\n",
    "    },\n",
    ")\n",
    "\n",
    "config_list_4v = autogen.config_list_from_json(\n",
    "    env_or_file=\"modelConfig.json\",\n",
    "    file_location=\".\",\n",
    "    filter_dict={\n",
    "        \"model\": [\"gpt-4-vision-preview\"],\n",
    "    },\n",
    ")\n",
    "\n",
    "assert len(config_list) > 0\n",
    "print(\"models to use: \", [config_list[i][\"model\"] for i in range(len(config_list))])\n",
    "\n",
    "assert len(config_list_4v) > 0\n",
    "print(\"vision models to use: \", [config_list_4v[i][\"model\"] for i in range(len(config_list))])\n",
    "\n",
    "# autogen.ChatCompletion.start_logging()\n",
    "termination_msg = lambda x: isinstance(x, dict) and \"TERMINATE\" == str(x.get(\"content\", \"\"))[-9:].upper()\n",
    "\n",
    "# define agents\n",
    "PI = autogen.UserProxyAgent(\n",
    "    name=\"PI\",\n",
    "    is_termination_msg=termination_msg,\n",
    "    human_input_mode=\"NEVER\",\n",
    "    system_message=\"Hi, You are the Principal Investigator of BioVerse. Welcome to the virtual lab! Your role is to identify promising research directions, define project goals, and keep the other members on track for delivering impactful results.\",\n",
    "    code_execution_config=False,  # we don't want to execute code in this case.\n",
    ")\n",
    "\n",
    "PI_aid = RetrieveUserProxyAgent(\n",
    "    name=\"PI_Assistant\",\n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Assistant who has extra content retrieval power for solving difficult problems.\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    max_consecutive_auto_reply=3,\n",
    "    retrieve_config={\n",
    "        \"task\": \"code\",\n",
    "        \"docs_path\": \"group_chat_pdf\",\n",
    "        \"chunk_token_size\": 1000,\n",
    "        \"model\": config_list[0][\"model\"],\n",
    "        \"client\": chromadb.PersistentClient(path=\"group_chat_db\"),\n",
    "        \"collection_name\": \"groupchat\",\n",
    "        \"get_or_create\": True,\n",
    "    },\n",
    "    code_execution_config=False,  # we don't want to execute code in this case.\n",
    ")\n",
    "\n",
    "critic = RetrieveAssistantAgent(\n",
    "    name=\"PI\",\n",
    "    system_message=\"Critic. Double check plan, claims, code from other agents and provide feedback. Check whether the plan includes adding verifiable info such as url of research paper. Please verify the links and execute the code if possible.\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\n",
    "        \"timeout\": 120,\n",
    "        \"config_list\": config_list,\n",
    "        \"temperature\": 0.5, \n",
    "        \"max_tokens\": 300\n",
    "    }\n",
    ")\n",
    "\n",
    "research_assistant_assistant = RetrieveAssistantAgent(\n",
    "    name=\"research_assistant_agent\",\n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Greetings, You are the research assistant. You excel at literature reviews, proposal preparations, quantitative and qualitative data collection/analysis, lab experiment assistance, and research project coordination. Please help me find relevant papers, take notes, format documents, track project timelines, locate lab equipment, ensure protocol adherence, or any other tasks that will accelerate my research!\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\n",
    "        \"timeout\": 120,\n",
    "        \"config_list\": config_list,\n",
    "        \"temperature\": 0.5, \n",
    "        \"max_tokens\": 300\n",
    "    }\n",
    ")\n",
    "\n",
    "image_analyst_assistant = RetrieveAssistantAgent(\n",
    "    name=\"image_analyst_agent\", \n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Hi there, You are the image analyst! Your specialty is processing and analyzing microscopic, radiographic, and other visual data. You can help with tasks like image segmentation, feature extraction, image registration, measuring objects in images, and classifying/annotating images using machine learning. Please provide assistance on analyzing, enhancing, or extracting meaningful information from my imaging data!\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\"timeout\": 120,\n",
    "                \"config_list\": config_list, \n",
    "                \"temperature\": 0.5, \n",
    "                \"max_tokens\": 300}\n",
    ")\n",
    "\n",
    "experimentalist_assistant = RetrieveAssistantAgent(\n",
    "    name=\"experimentalist_agent\", \n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Hello! You are the experimentalist. Your skills are in designing and conducting hands-on lab experiments, including following protocols, collecting and recording data, lab maintenance, and safe chemical/equipment handling. Please help me brainstorming experiments, setting up lab equipment, following detailed protocols, tracking measurements and observations, or assistance with other in-lab experimental work.\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\n",
    "        \"timeout\": 120,\n",
    "        \"config_list\": config_list,\n",
    "        \"temperature\": 0.5, \n",
    "        \"max_tokens\": 300\n",
    "    }\n",
    ")\n",
    "\n",
    "bioinformatics_assistant = RetrieveAssistantAgent(\n",
    "    name=\"bioinformatics_agent\", \n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Hello, you are the bioinformatician. You can help me with analyzing and interpreting biological data. Your skills include statistical analysis, DNA/RNA sequencing analysis, designing bioinformatics workflows, and providing guidance on best practices for computational biology research. Please help me analyze and make sense of large-scale omics data!\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\n",
    "        \"timeout\": 120,\n",
    "        \"config_list\": config_list,\n",
    "        \"temperature\": 0.5, \n",
    "        \"max_tokens\": 300\n",
    "    },\n",
    "    code_execution_config={\"last_n_messages\": 3, \"work_dir\": \"paper\"}\n",
    ")\n",
    "\n",
    "tissue_engineering_assistant = RetrieveAssistantAgent(\n",
    "    name=\"tissue_engineering_agent\", \n",
    "    is_termination_msg=termination_msg,\n",
    "    system_message=\"Hola, You are the tissue engineer here to help with designing and culturing organoid models. Your expertise includes selecting cell sources, extracellular matrix components, growth factors, and culture systems to derive organoid structures from stem cells. Please provide assistance planning organoid protocols, troubleshooting culture conditions, or analyzing resulting model viability and morphology. Please help me to advance my organ-on-a-chip and 3D tissue engineering goals!\",\n",
    "    human_input_mode=\"NEVER\",\n",
    "    llm_config={\n",
    "        \"timeout\": 120,\n",
    "        \"config_list\": config_list,\n",
    "        \"temperature\": 0.5, \n",
    "        \"max_tokens\": 300\n",
    "    }\n",
    ")\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T11:22:18.482712Z",
     "start_time": "2023-11-12T11:22:16.317888Z"
    }
   },
   "id": "4aae1f4bfd5d218d"
  },
  {
   "cell_type": "markdown",
   "source": [
    "Solve task"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5e4a67d463942177"
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "PROBLEM = \"Can you come up with a starting point for the problem posed by the PI regarding organoids with diagrams?\"\n",
    "\n",
    "llm_config = {\n",
    "    \"timeout\": 60,\n",
    "    \"seed\": 50,\n",
    "    \"config_list\": config_list,\n",
    "    \"temperature\": 0.5,\n",
    "}\n",
    "\n",
    "def _reset_agents():\n",
    "    PI.reset()\n",
    "    PI_aid.reset()\n",
    "    research_assistant_assistant.reset()\n",
    "    bioinformatics_assistant.reset()\n",
    "    experimentalist_assistant.reset()\n",
    "    image_analyst_assistant.reset()\n",
    "    critic.reset()\n",
    "\n",
    "def rag_chat():\n",
    "    _reset_agents()\n",
    "    groupchat = autogen.GroupChat(\n",
    "        agents=[PI_aid, research_assistant_assistant,bioinformatics_assistant, experimentalist_assistant, image_analyst_assistant, tissue_engineering_assistant, critic], messages=[], max_round=12\n",
    "    )\n",
    "    manager = autogen.GroupChatManager(groupchat=groupchat, llm_config=llm_config)\n",
    "\n",
    "    # Start chatting with boss_aid as this is the user proxy agent.\n",
    "    PI_aid.initiate_chat(\n",
    "        manager,\n",
    "        problem=PROBLEM,\n",
    "        n_results=3,\n",
    "    )\n",
    "\n",
    "def norag_chat():\n",
    "    _reset_agents()\n",
    "    groupchat = autogen.GroupChat(\n",
    "        agents=[PI_aid, research_assistant_assistant,bioinformatics_assistant, experimentalist_assistant, image_analyst_assistant, tissue_engineering_assistant, critic], messages=[], max_round=12\n",
    "    )\n",
    "    manager = autogen.GroupChatManager(groupchat=groupchat, llm_config=llm_config)\n",
    "\n",
    "    # Start chatting with boss as this is the user proxy agent.\n",
    "    PI.initiate_chat(\n",
    "        manager,\n",
    "        message=PROBLEM,\n",
    "    )\n",
    "\n",
    "def call_rag_chat():\n",
    "    _reset_agents()\n",
    "    # In this case, we will have multiple user proxy agents and we don't initiate the chat\n",
    "    # with RAG user proxy agent.\n",
    "    # In order to use RAG user proxy agent, we need to wrap RAG agents in a function and call\n",
    "    # it from other agents.\n",
    "    def retrieve_content(message, n_results=3):\n",
    "        PI_aid.n_results = n_results  # Set the number of results to be retrieved.\n",
    "        # Check if we need to update the context.\n",
    "        update_context_case1, update_context_case2 = PI_aid._check_update_context(message)\n",
    "        if (update_context_case1 or update_context_case2) and PI_aid.update_context:\n",
    "            PI_aid.problem = message if not hasattr(PI_aid, \"problem\") else PI_aid.problem\n",
    "            _, ret_msg = PI_aid._generate_retrieve_user_reply(message)\n",
    "        else:\n",
    "            ret_msg = PI_aid.generate_init_message(message, n_results=n_results)\n",
    "        return ret_msg if ret_msg else message\n",
    "    \n",
    "    PI_aid.human_input_mode = \"NEVER\" # Disable human input for boss_aid since it only retrieves content.\n",
    "    \n",
    "    llm_config = {\n",
    "        \"functions\": [\n",
    "            {\n",
    "                \"name\": \"retrieve_content\",\n",
    "                \"description\": \"retrieve content for code generation and question answering.\",\n",
    "                \"parameters\": {\n",
    "                    \"type\": \"object\",\n",
    "                    \"properties\": {\n",
    "                        \"message\": {\n",
    "                            \"type\": \"string\",\n",
    "                            \"description\": \"Refined message which keeps the original meaning and can be used to retrieve content for code generation and question answering.\",\n",
    "                        }\n",
    "                    },\n",
    "                    \"required\": [\"message\"],\n",
    "                },\n",
    "            },\n",
    "        ],\n",
    "        \"config_list\": config_list,\n",
    "        \"timeout\": 60,\n",
    "        \"seed\": 42,\n",
    "    }\n",
    "\n",
    "    for agent in [research_assistant_assistant,bioinformatics_assistant, experimentalist_assistant, image_analyst_assistant, tissue_engineering_assistant, critic]:\n",
    "        # update llm_config for assistant agents.\n",
    "        agent.llm_config.update(llm_config)\n",
    "\n",
    "    for agent in [PI, research_assistant_assistant,bioinformatics_assistant, experimentalist_assistant, image_analyst_assistant, tissue_engineering_assistant, critic]:\n",
    "        # register functions for all agents.\n",
    "        agent.register_function(\n",
    "            function_map={\n",
    "                \"retrieve_content\": retrieve_content,\n",
    "            }\n",
    "        )\n",
    "\n",
    "    groupchat = autogen.GroupChat(\n",
    "        agents=[PI, research_assistant_assistant,bioinformatics_assistant, experimentalist_assistant, image_analyst_assistant, tissue_engineering_assistant, critic], messages=[], max_round=12\n",
    "    )\n",
    "    manager = autogen.GroupChatManager(groupchat=groupchat, llm_config=llm_config)\n",
    "\n",
    "    # Start chatting with boss as this is the user proxy agent.\n",
    "    PI.initiate_chat(\n",
    "        manager,\n",
    "        message=PROBLEM,\n",
    "    )"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T11:22:22.004082Z",
     "start_time": "2023-11-12T11:22:22.001844Z"
    }
   },
   "id": "87a7811e395dbb4d"
  },
  {
   "cell_type": "markdown",
   "source": [
    "start chat"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "7575b5c7b09cde02"
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[33mPI\u001B[0m (to chat_manager):\n",
      "\n",
      "Can you come up with a starting point for the problem posed by the PI regarding organoids with diagrams?\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mresearch_assistant_agent\u001B[0m (to chat_manager):\n",
      "\n",
      "Certainly! To start addressing the problem posed by the PI regarding organoids, we can begin by conducting a literature review to understand the current state of research in organoid development, their applications, and any existing challenges. Additionally, creating diagrams to illustrate the structure and development of organoids can be helpful for visualizing their complexity and potential for modeling organ development and diseases.\n",
      "\n",
      "I can assist in conducting a literature review to gather relevant papers on organoid development and applications, create diagrams or visual aids to illustrate organoid structure and development, and help in formatting documents to present the gathered information in a clear and organized manner. Let me know how you would like to proceed!\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mPI_Assistant\u001B[0m (to chat_manager):\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mPI\u001B[0m (to chat_manager):\n",
      "\n",
      "To address the problem posed by the PI regarding organoids, I propose the following plan:\n",
      "\n",
      "1. Literature Review:\n",
      "   - Conduct a comprehensive literature review to gather information on the current state of research in organoid development, their applications, and any existing challenges.\n",
      "   - Identify key research papers, review articles, and scientific publications related to organoid development and applications.\n",
      "\n",
      "2. Diagram Creation:\n",
      "   - Create diagrams or visual aids to illustrate the structure and development of organoids.\n",
      "   - Visualize the complexity of organoids and their potential for modeling organ development and diseases.\n",
      "\n",
      "3. Document Formatting:\n",
      "   - Assist in formatting documents to present the gathered information, including literature review findings and diagrams, in a clear and organized manner.\n",
      "   - Ensure the inclusion of verifiable information such as the URLs of research papers and sources for transparency.\n",
      "\n",
      "Upon completion of the literature review, diagram creation, and document formatting, the gathered information can be presented to the PI for further discussion and potential incorporation into research proposals or presentations.\n",
      "\n",
      "I will begin by conducting the literature review and creating diagrams to provide a comprehensive overview of organoid research. Let me know if there are any specific areas of focus or additional requirements for the literature review and diagram creation.\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mresearch_assistant_agent\u001B[0m (to chat_manager):\n",
      "\n",
      "Great! I will start by conducting a comprehensive literature review to gather information on organoid development, applications, and challenges. I will identify key research papers, review articles, and scientific publications related to organoid development and applications. Additionally, I will create diagrams or visual aids to illustrate the structure and development of organoids. Finally, I will assist in formatting documents to present the gathered information in a clear and organized manner, ensuring the inclusion of verifiable information.\n",
      "\n",
      "I will keep you updated on the progress and findings as I work through the literature review and diagram creation. If there are specific journals, databases, or keywords you would like me to focus on during the literature review, please let me know.\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mPI_Assistant\u001B[0m (to chat_manager):\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\u001B[33mPI_Assistant\u001B[0m (to chat_manager):\n",
      "\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "norag_chat()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-11-12T11:22:36.991121Z",
     "start_time": "2023-11-12T11:22:25.144456Z"
    }
   },
   "id": "4b194c8910df5b6c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "8bff40128034f1dc"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
